---
title: "Getting Started"
author: "Connor H Knight"
date: "11/08/2021"
output: 
    html_document:
    toc: true
    toc_depth: 3
    theme: united
vignette: >
  %\VignetteIndexEntry{Getting Started}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r klippy, echo=FALSE, include=TRUE}
klippy::klippy(tooltip_message = 'Click to copy', tooltip_success = 'Done', position = c('bottom', 'right'))
```

# Analysing 2994 Bone Marrow Mononuclear Cells (BMMC) from a single sample.

To demonstrate IBRAP we enlisted a sample produced by [Oetjen](https://insight.jci.org/articles/view/124928) *et al*. This can be downloaded using the following link:  

https://www.dropbox.com/s/3wj4f48x7txuyrp/BMMC.rds?dl=0

```{r, eval=F, echo=T}

bmmc <- readRDS(file.choose())
  
                       
```

## Chapter 1: Cleaning droplet-based datasdets

Droplet-based technologies are normally accompanied by specific conundra that requires omission, namely: multiplets and ambient RNA. With IBRAP you can apply a python package called scrublet (perform.scrublet()) and decontX (perform.decontX()) from the celda package in R for removal, respectively. We supply our count matrix to the functions. There is no defined order of applciation, but logically we assume scrublet removal followed by decontX. 

```{r, eval=F, echo=T}

# We simply supply our generated count matrix to both functions

bmmc <- perform.scrublet(counts = bmmc)

bmmc <- perform.decontX(counts = bmmc)

```

```{r, echo=FALSE, out.width="150%", out.height="150%"}

knitr::include_graphics(path = 'scrublet_plot.png')

```

```{r, echo=FALSE, out.width="150%", out.height="150%"}

knitr::include_graphics(path = 'decontX.png')

```

## Chapter 2: Initiating the IBRAP S4 object

Since our datasets are produced by 10x genomics and aligned using CellRanger, 3 subsequent files are produced to integrate these files into a count matrix.

Now we can initiate our IBRAP class object ready for downstream analyses. Refer to methods-object and IBRAP-object for further detail. 

```{r, eval=F, echo=T}

# whilst creating the object we begin filtering the cells and genes using min.cells and min.features
bmmc <- createIBRAPobject(counts = bmmc,
                          original.project = 'bmmc',
                          method.name = 'RAW',
                          min.cells = 3,
                          min.features = 200)
                          
```

## Chapter 3: Quality Control

Current best practices for quality assuring scRNA-seq datasets are outlined in numerous [publications](https://www.embopress.org/doi/full/10.15252/msb.20188746). Including:

```{r table2, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
df <- data.frame(`QC Metric` = c('Mitochondrial Fraction', 'Total RNA', 'Total Genes'), Description = c('An abnormally high mitochondrail fraction indicates improper wet lab preparation', 'Low RNA quanitity indicates droplets containing no cells whereas, disproportionately high RNA indicates a droplet containing more than one cell', 'likewise, a low or high number of genes typically correspond with total RNA'))

knitr::kable(df)
```

The total RNA and Total genes are calculated upon the creation of the IBRAP object. However, percentage mitochondrial genes requires a second function since mitochondrial genes have a varying prefix in datasets. Thus, we use find_percentage_genes():

```{r, eval=F, echo=T}

bmmc <- find_percentage_genes(object = bmmc, pattern = '^MT-',
                              assay = 'RAW', 
                              slot = 'counts',
                              column.name = 'RAW_percent.mt')
                              
```

We can visualise this information in two ways: using violin plots (plot.qc.vln()) or comparing two metrices with scatter plots (plot.qc.scatter()). The latter enables us to observe correlations between the metrices. 

```{r, eval=F, echo=T}

plot.QC.vln(object = bmmc, 
            metadata.columns = c('RAW_total.features', 
                                 'RAW_total.counts', 
                                 'RAW_percent.mt'))
                                 
```

```{r, echo=FALSE, out.width="150%", out.height="150%"}

knitr::include_graphics(path = 'vln_qc_pre.png')

```

```{r, eval=F, echo=T}

plot.qc.scatter(object = bmmc, 
                y = 'RAW_total.counts', 
                x = 'RAW_percent.mt', 
                split.by = 'original.project'
                )

```

```{r, echo=FALSE, out.width="150%", out.height="150%"}

knitr::include_graphics(path = 'scatter_qc_pre.png')

```

Filtration is subsequently applied to remove problematic cells.

```{r, eval=F, echo=T}

bmmc <- filter_IBRAP(RAW_total.features < 2500 & RAW_total.counts > 200 & RAW_percent.mt < 8)

```

## Chapter 4: Data Transformation

To accommodate downstream analyses, we must promote a normal distribution and numerical scaling in our data. We included four separate [normalisation](https://connorhknight.github.io/IBRAP/articles/data_transform.html) methods that are applied as such:

```{r, eval=F, echo=T}

bmmc <- perform.sct(object = bmmc, 
                    assay = 'RAW', 
                    slot = 'counts')

bmmc <- perform.scran(object = bmmc, 
                      assay = 'RAW', 
                      slot = 'counts', 
                      vars.to.regress = 'RAW_total.counts', do.scale = T)

bmmc <- perform.scanpy(object = bmmc, 
                       vars.to.regress = 'RAW_total.counts', do.scale = T)

```

Please refer to the above hyperlink to understand why we only used 3 out of 4 methods for this 10x dataset. 

Each normalisation function produces a new method-assay which contains the results. Any downstream analyses will be stored in these separate compartments, respective to the normalisation method. i.e. object@methods contains all of the method assays. Hereon, function will iterate through the indicated method-assays and any other inputs within the assays. For example, if you supply c('SCT','SCRAN','SCANPY') as assays and c('pca_umap', 'pca_t-SNE') as reductions. A function will perform firstly on 'pca_umap' in the 'SCT' assay, then the 'pca_t-SNE' in the 'SCT' assay. After these the function will move on to 'pca_umap' in the 'SCRAN' assay and so on.

In some instances, certain genes require omission from the highly varaible genes. This can be achieved with the following function: 

```{r, eval=F, echo=T}

# This function finds all ribosomal genes and removes them from the HVG list of the three assays
unwanted <- as.character(rownames(bmmc)[grepl("^RPL" ,x=rownames(bmmc))])
bmmc <- remove.hvgs(object = bmmc, assay = c('SCT', 'SCRAN', 'SCANPY'), 
                    hvgs.omit = list(unwanted,unwanted,unwanted))

```

## Chapter 5: Linear Dimensionaly Reduction

Now that our data has been appropriately transformed we can reduce the number of dimensions using Principal Component Analysis (PCA) perform.pca(). Reduction methods summarise as much information from the higher dimensions into lower dimensions as possible. 

```{r, eval=F, echo=T}

bmmc <- perform.pca(object = bmmc, 
                    assay = c('SCT', 'SCRAN', 'SCANPY'), 
                    n.pcs = 50, reduction.save = 'pca')

```

The produce principal components (PCs) contain a different explained variance, high variance PCs contain the most information between points and low provide the opposite. Furthermore, low variation PCs create a false closeness between cell points and thus require exclusion from *some* downstream analyses. A simple way to identify the number of PCs to retain is to plot them in descending order and find the elbow point. The elbow point being the bend in the plot of points. 

```{r, echo=FALSE, out.width="150%", out.height="150%"}

knitr::include_graphics(path = 'PCA_screeplot.png')

```

For these assays, the elbow point is around 10-12 PCs. 

## Chapter 6: Generating Neighbourhood Graphs

IBRAP enlists two neighbourhood graph generators: 
 - perform.nn.v1() which is derived scanpys neighbourhood graph function.
 - perform.nn.v2() which is seurats version. 
 
The main difference is that v1 uses UMAPs method of finding neighbours and all edge weights equate to one. Whereas, v2 uses the Louvain algorithm to maximise modularity (network closeness) and edge weights are jacquard similarity between neighbours. 

Scanpy also integrated an ease-of-use function to generate diffusion maps generated from the neighbourhood graph. This is an equivalent to PCA but is thought to represent cellular trajectories. We can generate these diffusion maps by setting the generate.diffmap parameter to TRUE. Since we need to generate neighourhood graphs for the diffusion maps too, there is a little repetition in functon application.

```{r, eval=F, echo=T}
# here we generate the neighbour graph for PCA and produce a diffusion map from the PCA-neighbour graph

bmmc <- perform.nn.v1(object = bmmc, assay = c('SCT', 'SCRAN', 'SCANPY'), 
                      reduction = c('pca'), 
                      dims = list(0), 
                      generate.diffmap = T)
                      
# the next function generates neighbours for the diffusion map
                      
bmmc <- perform.nn.v1(object = bmmc, 
                      assay = c('SCT', 'SCRAN', 'SCANPY'), 
                      reduction = c('pca_nn.v1:diffmap'), 
                      dims = list(0),
                      generate.diffmap = F)
                      
# this last function finds the neighbour graph for pca and the difusion map using seurats version
# v2 does not produce diffusion maps

bmmc <- perform.nn.v2(object = bmmc, 
                      assay = c('SCT', 'SCRAN', 'SCANPY'), 
                      reduction = c('pca', 'pca_nn.v1:diffmap'), 
                      dims = list(0,0))
                          
```

## Chapter 7: Non-linear Reduction

Since we have reduced our dataset into a manageable size, produce neighbourhood graphs and clustered communities, we need to visualise our data. scRNA-seq contains many zeros which makes a sparse matrix, this renders linear reduction methods inappropriate for visualisation. Therefore, we resort to non-linear reductions.

IBRAP integrated the ability to perform 3 methods:
- UMAP
- t-SNE
- lvish (a LargeVis adaptation)

For this tutorial we will focus on UMAP. 

```{r, eval=F, echo=T}

bmmc <- perform.umap(object = samp, 
                     assay = c('SCT', 'SCRAN', 'SCANPY'), 
                     reduction = c('pca', 'pca_nn.v1:diffmap'), 
                     n.dims = list(1:10,NULL))

```

## Chapter 8: Clustering

Finally, using our neighbourhood graph we must cluster close networks together to infer individual cell types. perform.graph.cluster() provides either:
- Louvain algorithm (1)
- Louvain algorithm with multilevel refinement (2)
- Smart Local Moving (SLM) (3)
- Leiden (4)

```{r, eval=F, echo=T}

bmmc <- perform.graph.cluster(object = bmmc, 
                               assay = c('SCT', 'SCRAN', 'SCANPY'), 
                               neighbours = c("pca", "pca_nn.v1:diffmap"), 
                               algorithm = 1)

```

There are alternative [clustering methods](https://connorhknight.github.io/IBRAP/articles/clustering_method.html) that are described in an alternate article


After generating reductions and cell assignments you may wish to benchmark these assignments, this is demonstrated in this tutorial

## Chapter 9: Visualising Clustering Results

Since IBRAP has integrated several pipelines we must begin to visualise the differences between results. 

We performed Louvain graph clustering from two different neighbourhood graphs which can causes varied results. Therefore we should observe the difference between these values. 

```{r, eval=F, echo=T}

plot1 <- plot.reduced.dim(object = samp, 
                          reduction = 'pca_umap', 
                          assay = 'SCT', 
                          clust.method = 'pca_nn.v1:louvain', 
                          column = 'neighbourhood_graph_res.0.7', pt.size = 0.1) + 
  ggplot2::ggtitle('neighbour_v1') + 
  ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5))

plot2 <- plot.reduced.dim(object = samp, 
                          reduction = 'pca_umap', 
                          assay = 'SCT', 
                          clust.method = 'pca_nn.v2:louvain', 
                          column = 'neighbourhood_graph_res.0.6', pt.size = 0.1) + 
  ggplot2::ggtitle('neighbour_v2') + 
  ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5))

plot1 + plot2

```

```{r, echo=FALSE, out.width="150%", out.height="150%"}

knitr::include_graphics(path = 'umap_plots_neighbours.png')

```

We also produce these results using different normalisation methods, lets observe the difference between these. 

```{r, eval=F, echo=T}

plot3 <- plot.reduced.dim(object = samp, 
                          reduction = 'pca_umap', 
                          assay = 'SCT', 
                          clust.method = 'pca_nn.v1:louvain', 
                          column = 'neighbourhood_graph_res.0.7', pt.size = 0.1) + 
  ggplot2::ggtitle('SCT_neighbour_v1') + 
  ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5))

plot4 <- plot.reduced.dim(object = samp, 
                          reduction = 'pca_umap', 
                          assay = 'SCRAN', 
                          clust.method = 'pca_nn.v1:louvain', 
                          column = 'neighbourhood_graph_res.0.7', pt.size = 0.1) + 
  ggplot2::ggtitle('SCRAN_neighbour_v1') + 
  ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5))

plot5 <- plot.reduced.dim(object = samp, 
                          reduction = 'pca_umap', 
                          assay = 'SCANPY', 
                          clust.method = 'pca_nn.v1:louvain', 
                          column = 'neighbourhood_graph_res.0.7', pt.size = 0.1) + 
  ggplot2::ggtitle('SCANPY_neighbour_v1') + 
  ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5))

plot3 + plot4 + plot5

```

```{r, echo=FALSE, out.width="150%", out.height="150%"}

knitr::include_graphics(path = 'diff_norm_method.png')

```

We shall proceed using the SCT neighbourhood graph version 1. We have clustered our cells however, now we must identify the biology. 

## Chapter 10: Benchmarking Clustering Results

A large number of results will be produced using IBRAP therefore it is beneficial to observe the effectiveness of the clustering results. IBRAP contains a benchmarking function that enlists multiple metrices to assess clustering performance. These metrices are only informative however and will be unable to distinguish the best results (that is down to the user to decide). 

To set up the function, supply the dataframes contained within the @cluster_assignments you wish to benchmark under clustering. Next, include the corresponding reductions for them, normally this will be a umap reduction. For example, if clustering and the umap were calculated using PCA these should be paired. 

```{r, eval=F, echo=T}

samp <- benchmark.clustering(object = samp, assay = c('SCT', 'SCRAN', 'SCANPY'), 
                             clustering = c("pca_nn.v1:louvain","pca_nn.v1:diffmap_nn.v1:louvain",
                                            "pca_nn.v2:louvain","pca_nn.v1:diffmap_nn.v2:louvain"), 
                             reduction = c('pca_umap', 'pca_nn.v1:diffmap_umap',
                                           'pca_umap', 'pca_nn.v1:diffmap_umap'
                             ), 
                             n.dims = 1:2)

```

## Chapter 11: Biological Analysis 

To understand the biology driving the clustering we can tackle it in two ways: differential expression analyses or visualisation

Differential expression analyses can be performed either 1 cluster compared against all other cells or 1 cluster against 1 other cluster. In this tutorial we will use the 1 v all function. 

```{r, eval=F, echo=T}

DE_res <- perform.diffexp.all(object = samp, 
                              assay = 'SCT', 
                              test = 'MAST', 
                              identity = as.vector(samp@methods$SCT@cluster_assignments$`pca_nn.v1:louvain`$neighbourhood_graph_res.0.7))

```

These results then enable us to distinguish the unique markers that are bing expressed by each cluster. 

If the driving biology remains unclear from a differential expression analysis we may utilise visualisation. We have included 4 different methods of visualisation: feature plots, violin plots, dot plots, and heatmaps. 

In this case, we can use feature plots:

```{r, eval=F, echo=T}

plot.features(object = samp, assay = 'SCT', reduction = 'pca_umap', 
              features = c('JCHAIN','MS4A1','MME','IRF8','HLA-DRA','LYZ',
                           'ELANE','ALAS2','CD34','NKG7','CCR7','IL7R','GZMK'), 
              pt_size = 1, 
              slot = 'normalised')

```

```{r, echo=FALSE, out.width="150%", out.height="150%"}

knitr::include_graphics(path = 'Feature_plots.png')

```

Next, we must indicate which cell type belongs to which clusters

```{r, eval=F, echo=T}

clusters[clusters == '18'] <- 'plasma'
clusters[clusters == '17'] <- 'immature B cells'
clusters[clusters == '11'] <- 'pro B cells'
clusters[clusters == '14'] <- 'dendritic cells'
clusters[clusters == '15'] <- 'dendritic cells'
clusters[clusters == '1'] = 'monocytes/macrophage'
clusters[clusters == '10'] = 'monocytes/macrophage'
clusters[clusters == '6'] = 'neutrophil'
clusters[clusters == '5'] = 'erythrocytes'
clusters[clusters == '8'] = 'erythrocytes'
clusters[clusters == '9'] = 'erythrocytes'
clusters[clusters == '12'] = 'erythrocytes'
clusters[clusters == '16'] = 'erythrocytes'
clusters[clusters == '3'] = 'HSPCs'
clusters[clusters == '4'] = 'NK cells'
clusters[clusters == '0'] = 'CD4+ T cells'
clusters[clusters == '2'] = 'CD8+ naive T cells'
clusters[clusters == '13'] = 'CD8+ naive T cells'
clusters[clusters == '7'] = 'CD8+ cytotoxic T cells'

samp@sample_metadata$celltypes <- clusters

plot.reduced.dim(object = samp, 
                 reduction = 'pca_umap', 
                 assay = 'SCT', 
                 clust.method = 'metadata', 
                 column = 'celltypes')

```

```{r, echo=FALSE, out.width="150%", out.height="150%"}

knitr::include_graphics(path = 'cell_types.png')

```

