<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>GSEA â€¢ IBRAP</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="GSEA">
<meta property="og:description" content="IBRAP">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">IBRAP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Installation_linux.html">Installation (linux)</a>
    </li>
    <li>
      <a href="../articles/installation_macos.html">Installation (macos)</a>
    </li>
    <li>
      <a href="../articles/dockers.html">Using Docker/Singularity</a>
    </li>
    <li>
      <a href="../articles/using_IBRAP.html">Getting Started</a>
    </li>
    <li>
      <a href="../articles/integration.html">Data Integration</a>
    </li>
    <li>
      <a href="../articles/application.html">RShiny Application</a>
    </li>
    <li>
      <a href="../articles/trajectory.html">Trajectory Inference</a>
    </li>
    <li>
      <a href="../articles/singleR_annotation.html">Automated Annotation</a>
    </li>
    <li>
      <a href="../articles/data_transform.html">Data Transformation</a>
    </li>
    <li>
      <a href="../articles/clustering_method.html">Clustering Methods</a>
    </li>
    <li>
      <a href="../articles/HPC_issues.html">HPC issues/solutions</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="GSEA_files/clipboard-1.7.1/clipboard.min.js"></script><link href="GSEA_files/primer-tooltips-1.4.0/build.css" rel="stylesheet">
<link href="GSEA_files/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet">
<script src="GSEA_files/klippy-0.0.0.9500/js/klippy.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>GSEA</h1>
                        <h4 data-toc-skip class="author">Connor H Knight</h4>
            
            <h4 data-toc-skip class="date">11/08/2021</h4>
      
      
      <div class="hidden name"><code>GSEA.Rmd</code></div>

    </div>

    
    
<script>
  addClassKlippyTo("pre.r, pre.markdown");
  addKlippy('right', 'bottom', 'auto', '1', 'Click to copy', 'Done');
</script><div class="section level2">
<h2 id="perfoming-gene-set-enricvhment-with-ssgsea">Perfoming gene set enricvhment with ssGSEA<a class="anchor" aria-label="anchor" href="#perfoming-gene-set-enricvhment-with-ssgsea"></a>
</h2>
<p>Gene Set Enrichment Analysis enables us to determine if a defined set of genes shows a statistically difference between biological states. Here we apply this on single-cell RNA-seq data. In this tutorial, we show you how to enrich certain pathways in your single-cell data.</p>
<div class="section level3">
<h3 id="running-gsea">Running GSEA<a class="anchor" aria-label="anchor" href="#running-gsea"></a>
</h3>
<p>We begin by loading some test data: This command is simple loading in some example data and inserting it into an IBRAP object.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">IBRAP</span><span class="op">)</span></span>
<span><span class="co">#&gt; Welcome to IBRAP</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">'curl -LJO https://raw.githubusercontent.com/connorhknight/IBRAP/main/data/celseq2.rds'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">celseq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'celseq2.rds'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createIBRAPobject.html">createIBRAPobject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">celseq2</span><span class="op">$</span><span class="va">counts</span>, original.project <span class="op">=</span> <span class="st">'celseq2'</span>, meta.data <span class="op">=</span> <span class="va">celseq2</span><span class="op">$</span><span class="va">metadata</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.sct.html">perform.sct</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span><span class="op">)</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.pca.html">perform.pca</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, assay <span class="op">=</span> <span class="st">'SCT'</span><span class="op">)</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.umap.html">perform.umap</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, assay <span class="op">=</span> <span class="st">'SCT'</span>, reduction <span class="op">=</span> <span class="st">'PCA'</span>, dims.use <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">20</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We next next need to gather gene sets:</p>
<p>There are a couple of ways to do this: (1) get the gene sets from the escape package or (2) define your own.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Option 1</span></span>
<span></span>
<span><span class="va">gene.sets</span> <span class="op">&lt;-</span> <span class="fu">escape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/escape/man/getGeneSets.html" class="external-link">getGeneSets</a></span><span class="op">(</span>library <span class="op">=</span> <span class="st">'H'</span>, species <span class="op">=</span> <span class="st">'Homo sapiens'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Option 2</span></span>
<span></span>
<span><span class="va">gene.sets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Tcell_signature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD2"</span>,<span class="st">"CD3E"</span>,<span class="st">"CD3D"</span><span class="op">)</span>,</span>
<span>                  Myeloid_signature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SPI1"</span>,<span class="st">"FCER1G"</span>,<span class="st">"CSF1R"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">gene.sets</span> <span class="op">&lt;-</span> <span class="fu">escape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/escape/man/getGeneSets.html" class="external-link">getGeneSets</a></span><span class="op">(</span>library <span class="op">=</span> <span class="st">'H'</span>, species <span class="op">=</span> <span class="st">'Homo sapiens'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># You can adjust the number of computing cores to dedicate to the calculations throguh the core parameter.</span></span>
<span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.gsea.html">perform.gsea</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, gene.sets <span class="op">=</span> <span class="va">gene.sets</span>, return_object <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Using sets of 1000 cells. Running 3 times."</span></span>
<span><span class="co">#&gt; Setting parallel calculations through a SnowParam back-end</span></span>
<span><span class="co">#&gt; with workers=6 and tasks=100.</span></span>
<span><span class="co">#&gt; Estimating ssGSEA scores for 50 gene sets.</span></span>
<span><span class="co">#&gt; Setting parallel calculations through a SnowParam back-end</span></span>
<span><span class="co">#&gt; with workers=6 and tasks=100.</span></span>
<span><span class="co">#&gt; Estimating ssGSEA scores for 50 gene sets.</span></span>
<span><span class="co">#&gt; Setting parallel calculations through a SnowParam back-end</span></span>
<span><span class="co">#&gt; with workers=6 and tasks=100.</span></span>
<span><span class="co">#&gt; Estimating ssGSEA scores for 50 gene sets.</span></span>
<span></span>
<span><span class="co"># We can then bind the results into our IBRAP object metadata.</span></span>
<span></span>
<span><span class="va">obj</span><span class="op">@</span><span class="va">sample_metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">obj</span><span class="op">@</span><span class="va">sample_metadata</span>, <span class="va">results</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualisations">Visualisations<a class="anchor" aria-label="anchor" href="#visualisations"></a>
</h3>
<p>Here we can utilise the DittoSeq package to easily visualise our metaprogrammes.</p>
</div>
<div class="section level3">
<h3 id="heatmaps">Heatmaps<a class="anchor" aria-label="anchor" href="#heatmaps"></a>
</h3>
<p>Firstly, we convert our IBRAP object into a SCE object to be compatible with dittoseq.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">dittoSeq</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: ggplot2</span></span>
<span></span>
<span><span class="va">dittoseq_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare.for.dittoSeq.html">prepare.for.dittoSeq</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, assay <span class="op">=</span> <span class="st">'SCT'</span>, slot <span class="op">=</span> <span class="st">'normalised'</span>, reduction <span class="op">=</span> <span class="st">'PCA_UMAP'</span>, clust.method <span class="op">=</span> <span class="st">'metadata'</span><span class="op">)</span></span></code></pre></div>
<p>Then, we plot our metaprogrammes in a heatmap.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoHeatmap.html" class="external-link">dittoHeatmap</a></span><span class="op">(</span><span class="va">dittoseq_obj</span>, genes <span class="op">=</span> <span class="cn">NULL</span>, metas <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">results</span><span class="op">)</span>, </span>
<span>             annot.by <span class="op">=</span> <span class="st">"celltype"</span>, </span>
<span>             fontsize <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span></code></pre></div>
<p><img src="GSEA_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>There are many other plots that can be performed on enrichment analysis which can be found in the following tutorial: <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/escape/inst/doc/vignette.html" class="external-link uri">http://www.bioconductor.org/packages/release/bioc/vignettes/escape/inst/doc/vignette.html</a></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Connor Knight, Faraz Khan.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
