<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Data Integration â€¢ IBRAP</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Data Integration">
<meta property="og:description" content="IBRAP">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">IBRAP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Installation_linux.html">Installation (linux)</a>
    </li>
    <li>
      <a href="../articles/installation_macos.html">Installation (macos)</a>
    </li>
    <li>
      <a href="../articles/dockers.html">Using Docker/Singularity</a>
    </li>
    <li>
      <a href="../articles/using_IBRAP.html">Getting Started</a>
    </li>
    <li>
      <a href="../articles/integration.html">Data Integration</a>
    </li>
    <li>
      <a href="../articles/application.html">RShiny Application</a>
    </li>
    <li>
      <a href="../articles/trajectory.html">Trajectory Inference</a>
    </li>
    <li>
      <a href="../articles/singleR_annotation.html">Automated Annotation</a>
    </li>
    <li>
      <a href="../articles/data_transform.html">Data Transformation</a>
    </li>
    <li>
      <a href="../articles/clustering_method.html">Clustering Methods</a>
    </li>
    <li>
      <a href="../articles/HPC_issues.html">HPC issues/solutions</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="integration_files/clipboard-1.7.1/clipboard.min.js"></script><link href="integration_files/primer-tooltips-1.4.0/build.css" rel="stylesheet">
<link href="integration_files/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet">
<script src="integration_files/klippy-0.0.0.9500/js/klippy.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Data Integration</h1>
                        <h4 data-toc-skip class="author">Connor H Knight
&amp; Faraz Khan</h4>
            
            <h4 data-toc-skip class="date">11/08/2021</h4>
      
      
      <div class="hidden name"><code>integration.Rmd</code></div>

    </div>

    
    
<script>
  addClassKlippyTo("pre.r, pre.markdown");
  addKlippy('right', 'bottom', 'auto', '1', 'Click to copy', 'Done');
</script><p>It is common in single-cell analyses to integrate multiple samples.
There are numerous package available for do this, however they function
differently and generate different results.Therefore, it is becoming
necessary to benchmark a few options before deciding on a method. For
this, IBRAP contains 4 integration methods and benchmarking capabilities
to distinguish the best solution.</p>
<div class="section level2">
<h2 id="integrating-multiple-datasets">Integrating Multiple Datasets<a class="anchor" aria-label="anchor" href="#integrating-multiple-datasets"></a>
</h2>
<p>For this tutorial, we have supplied two well-characterised pancreas
samples: CELseq2 and Smartseq2 which contain 2285 and 2394 cells,
respectively.</p>
<p>You can download the data for this tutorial using the following
links:</p>
<p><a href="https://github.com/connorhknight/IBRAP/raw/main/data/celseq2.rds" class="external-link">celseq2</a>
<a href="https://github.com/connorhknight/IBRAP/raw/main/data/smartseq2.rds" class="external-link">smartseq2</a></p>
<div class="section level3">
<h3 id="creating-ibrap-objects">Creating IBRAP objects<a class="anchor" aria-label="anchor" href="#creating-ibrap-objects"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Here we set the maximum amount of RAM R may use, for this we have allowed 4GB</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">IBRAP</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>future.globals.maxSize <span class="op">=</span> <span class="fl">12000</span> <span class="op">*</span> <span class="fl">1024</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">celseq2_items</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'~/path/to/celseq2/celseq2.rds'</span><span class="op">)</span></span>
<span><span class="va">smartseq2_items</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'~/path/to/celseq2/celseq2.rds'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">celseq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createIBRAPobject.html">createIBRAPobject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">celseq2_items</span><span class="op">$</span><span class="va">counts</span>,</span>
<span>                             meta.data <span class="op">=</span> <span class="va">celseq2_items</span><span class="op">$</span><span class="va">metadata</span>,</span>
<span>                             original.project <span class="op">=</span> <span class="st">'celseq2'</span>,</span>
<span>                             min.cells <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                             min.features <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span></span>
<span><span class="va">smartseq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createIBRAPobject.html">createIBRAPobject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">smartseq2_items</span><span class="op">$</span><span class="va">counts</span>, </span>
<span>                               meta.data <span class="op">=</span> <span class="va">smartseq2_items</span><span class="op">$</span><span class="va">metadata</span>,</span>
<span>                               original.project <span class="op">=</span> <span class="st">'smartseq2'</span>,</span>
<span>                               min.cells <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                               min.features <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="merging-ibrap-objects">Merging IBRAP objects<a class="anchor" aria-label="anchor" href="#merging-ibrap-objects"></a>
</h3>
<p>Datasets are then merged together. WARNING: Datasets can only be
merged straight after their creation with no downstream results.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/merge.html">merge</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">celseq2</span>, y <span class="op">=</span> <span class="va">smartseq2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="pre-processing-batch-visualisation">Pre-processing &amp; Batch Visualisation<a class="anchor" aria-label="anchor" href="#pre-processing-batch-visualisation"></a>
</h3>
<p>We then continue our analysis as normal up until PCA reduction:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.sct.html">perform.sct</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, </span>
<span>                        assay <span class="op">=</span> <span class="st">'RAW'</span>, </span>
<span>                        slot <span class="op">=</span> <span class="st">'counts'</span>, </span>
<span>                         verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.scran.html">perform.scran</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, </span>
<span>                          assay <span class="op">=</span> <span class="st">'RAW'</span>, </span>
<span>                          slot <span class="op">=</span> <span class="st">'counts'</span>, </span>
<span>                          vars.to.regress <span class="op">=</span> <span class="st">'RAW_total.counts'</span>, </span>
<span>                          verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.scanpy.html">perform.scanpy</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, </span>
<span>                           vars.to.regress <span class="op">=</span> <span class="st">'RAW_total.counts'</span>, </span>
<span>                           verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.pca.html">perform.pca</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, </span>
<span>                        assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, </span>
<span>                        n.pcs <span class="op">=</span> <span class="fl">50</span>, </span>
<span>                        reduction.save <span class="op">=</span> <span class="st">'PCA'</span>, </span>
<span>                        verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.umap.html">perform.umap</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, </span>
<span>                         assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, </span>
<span>                         reduction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'PCA'</span><span class="op">)</span>, </span>
<span>                         n_components <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                         verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/plot.variance.html">plot.variance</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, reduction <span class="op">=</span> <span class="st">'PCA'</span><span class="op">)</span></span></code></pre></div>
<p><img src="integration_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>Here we can see that the variance for PCA begins to elbow at around
10-15 PCs, for this tutorial we will use 20 PCs to capture maximum
information downstream.</p>
<p>Please read each integration method carefully since they all have
different requirements and downstream analyses.</p>
<p>Now that we have pre-processed our data, lets take a look at the
un-integrated batches:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'PCA_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCT'</span>,</span>
<span>                          clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCT'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'PCA_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCRAN'</span>,</span>
<span>                          clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCRAN'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">plot3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'PCA_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCANPY'</span>,</span>
<span>                          clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCANPY'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">egg</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/egg/man/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plots <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">plot1</span>, <span class="va">plot2</span>, <span class="va">plot3</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span>                        </span></code></pre></div>
<p><img src="integration_files/figure-html/unnamed-chunk-7-1.png" width="700"><img src="integration_files/figure-html/unnamed-chunk-8-1.png" width="700"><img src="integration_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>As we can see there are profound differences between the samples due
to technical variation. This requires correction.</p>
</div>
<div class="section level3">
<h3 id="integration-techniques">Integration Techniques<a class="anchor" aria-label="anchor" href="#integration-techniques"></a>
</h3>
<p>We have included 4 integration techniques:</p>
<ul>
<li>Seurat CCA (counts)</li>
<li>Scanorama (counts)</li>
<li>BBKNN (reduction)</li>
<li>Harmony (reduction)</li>
</ul>
<p>Some integration methods either correct the normalised counts
matrices or the reduction components.</p>
</div>
<div class="section level3">
<h3 id="using-harmony-scanorama-and-seurat-cca">Using Harmony, Scanorama, and Seurat CCA<a class="anchor" aria-label="anchor" href="#using-harmony-scanorama-and-seurat-cca"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.harmony.html">perform.harmony</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, </span>
<span>                            assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCRAN'</span>, <span class="st">'SCT'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, </span>
<span>                            batch <span class="op">=</span> <span class="st">'original.project'</span>, </span>
<span>                            reduction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'PCA'</span><span class="op">)</span>,  </span>
<span>                            max.iter.harmony <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                            dims.use <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">20</span><span class="op">)</span>, </span>
<span>                            verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Here we are performing Harmony on all PCA reductions derived from SCT, SCRAN, and SCANPY normalised </span></span>
<span><span class="co"># matrices. We have set the dims.use to 20 which means to use only 20 PCs from the PCA. We also </span></span>
<span><span class="co"># increased the max.iter.harmony to 100 from the default 10, to allow further correction of the data.</span></span>
<span>                            </span></code></pre></div>
<p>Harmony accepts reduced embeddings and produces a new reduced
embedding called harmony reductions. These reductions can then be used
to produce nearest neighbourhood graphs, any clustering method, and any
reduction method. Harmony functions well since you can adjust the
aggressiveness of the correction per co-variate that you specify using
the theta parameter.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.scanorama.html">perform.scanorama</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, </span>
<span>                              assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, </span>
<span>                              slot <span class="op">=</span> <span class="st">'norm.scaled'</span>, </span>
<span>                              batch <span class="op">=</span> <span class="st">'original.project'</span>, </span>
<span>                              verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># For this tutorial we used Scanoramas default parameters and applied to the normalisation using the</span></span>
<span><span class="co"># normalised and scaled matrix which is recommended by the developers, however you may use any of the # matrices for your sample if you wish, </span></span>
<span><span class="co"># i.e. raw counts or just normalised data. Scanorama creates its own reduction with a default of 50, you can re-apply the analysis to generate</span></span>
<span><span class="co"># fewer reductions if your results do not seem right. </span></span></code></pre></div>
<p>Rather than functioning on reduced embeddings, scanorama uses whole
data or featured-subsetted data matrices and stitches together the
batches in order of similarity. Scanorama produces a reduced embedding
that can be used for nearest neighbourhood graph generation, any
clustering method, and non-linear reduction method.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">pancreas_split</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/splitIBRAP.html">splitIBRAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, </span>
<span>                             split.by <span class="op">=</span> <span class="st">'original.project'</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">x</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">pancreas_split</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  </span>
<span>  <span class="va">pancreas_split</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.sct.html">perform.sct</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas_split</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                                     verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span>  <span class="va">pancreas_split</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.scran.html">perform.scran</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas_split</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                                       vars.to.regress <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'RAW_total.counts'</span><span class="op">)</span>, </span>
<span>                                       verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span>  <span class="va">pancreas_split</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.scanpy.html">perform.scanpy</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas_split</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                                        vars.to.regress <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'RAW_total.counts'</span><span class="op">)</span>, </span>
<span>                                        verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span>  </span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.seurat.cca.html">perform.seurat.integration</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, </span>
<span>                                       object.list <span class="op">=</span> <span class="va">pancreas_split</span>, </span>
<span>                                       assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>,<span class="st">'SCRAN'</span>,<span class="st">'SCANPY'</span><span class="op">)</span>, </span>
<span>                                       verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot.variance.html">plot.variance</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, </span>
<span>              assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, </span>
<span>              reduction <span class="op">=</span> <span class="st">'CCA'</span>, </span>
<span>              verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p>We next used Seurats CCA integration analysis to integrate our data.
This method finds feature anchors as the basis of the integration. This
method requires our batches to be separated and normalised individually
prior to integration. This is performed using the splitIBRAP function
and then looping through the batches and normalising them. The corrected
matrix is the reduced with a PCA and is stored under
integration_reductions as CCA however, this name can be changed in the
parameters.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Here we have listed the n.dims in the same order as they are presented in the reduction parameter, remember 0 is equal to all dimensions and</span></span>
<span><span class="co"># that each reduction may require different parameters. For example, we can see that the variance explained in the PCs is reduced to prior to 10</span></span>
<span><span class="co"># compared to the uncorrected PCA which had more variance spread across more PCs</span></span>
<span>                              </span>
<span><span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.umap.html">perform.umap</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, </span>
<span>                         assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, </span>
<span>                         reduction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'PCA_HARMONY'</span>, <span class="st">'SCANORAMA'</span>, <span class="st">'CCA'</span><span class="op">)</span>, </span>
<span>                         n_components <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                         n.dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">15</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.nn.html">perform.nn</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>,</span>
<span>                       assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, </span>
<span>                       reduction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'PCA_HARMONY'</span>, <span class="st">'SCANORAMA'</span>, <span class="st">'CCA'</span><span class="op">)</span>, </span>
<span>                       dims.use <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">15</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.graph.cluster.html">perform.graph.cluster</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, </span>
<span>                                  assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, </span>
<span>                                  neighbours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'PCA_HARMONY_NN'</span>, <span class="st">'SCANORAMA_NN'</span>, <span class="st">'CCA_NN'</span><span class="op">)</span>, </span>
<span>                                  algorithm <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="using-bbknn">Using BBKNN<a class="anchor" aria-label="anchor" href="#using-bbknn"></a>
</h3>
<p>BBKNN functions differently to the others, it does not produce a
reduced dimension but produces nearest neighbour graph and thus must be
treated slighlty differently downstream.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.bbknn.html">perform.bbknn</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, </span>
<span>                          assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, </span>
<span>                          reduction <span class="op">=</span> <span class="st">'PCA'</span>,</span>
<span>                          batch <span class="op">=</span> <span class="st">'original.project'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># For this part we used BBKNNs default parameters and applied to the PCA results we generated earlier.</span></span>
<span></span>
<span><span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.umap.html">perform.umap</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, </span>
<span>                         assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, </span>
<span>                         graph <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'PCA_BBKNN_BBKNN'</span><span class="op">)</span>, </span>
<span>                         n_components <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.graph.cluster.html">perform.graph.cluster</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, </span>
<span>                                  assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, </span>
<span>                                  neighbours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'PCA_BBKNN_BBKNN'</span><span class="op">)</span>, </span>
<span>                                  algorithm <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>We next created UMAP reduced embeddings from the previously generated
corrected reduced embeddings, including: PCA_HARMONY and SCANORAMA.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">plot.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot.list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'PCA_HARMONY_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCT'</span>,</span>
<span>                                   clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCT_harmony'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot.list</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'PCA_HARMONY_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCRAN'</span>,</span>
<span>                                   clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCRAN_harmony'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot.list</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'PCA_HARMONY_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCANPY'</span>,</span>
<span>                                   clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCANPY_harmony'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot.list</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'SCANORAMA_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCT'</span>,</span>
<span>                                   clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCT_SCANORAMA'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot.list</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'SCANORAMA_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCRAN'</span>,</span>
<span>                                   clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCRAN_SCANORAMA'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot.list</span><span class="op">[[</span><span class="fl">6</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'SCANORAMA_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCANPY'</span>,</span>
<span>                                   clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCANPY_SCANORAMA'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot.list</span><span class="op">[[</span><span class="fl">7</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'PCA_BBKNN_BBKNN:UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCT'</span>,</span>
<span>                                   clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCT_SCANORAMA'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot.list</span><span class="op">[[</span><span class="fl">8</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'PCA_BBKNN_BBKNN:UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCRAN'</span>,</span>
<span>                                   clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCRAN_SCANORAMA'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot.list</span><span class="op">[[</span><span class="fl">9</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'PCA_BBKNN_BBKNN:UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCANPY'</span>,</span>
<span>                                   clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCANPY_SCANORAMA'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot.list</span><span class="op">[[</span><span class="fl">10</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'CCA_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCT'</span>,</span>
<span>                                   clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCT_SCANORAMA'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot.list</span><span class="op">[[</span><span class="fl">11</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'CCA_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCRAN'</span>,</span>
<span>                                   clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCRAN_SCANORAMA'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot.list</span><span class="op">[[</span><span class="fl">12</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'CCA_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCANPY'</span>,</span>
<span>                                   clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCANPY_SCANORAMA'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">egg</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/egg/man/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plots <span class="op">=</span> <span class="va">plot.list</span>, nrow <span class="op">=</span> <span class="fl">4</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="integration_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p><img src="integration_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<p><img src="integration_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<p><img src="integration_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<p>From initial inspection we can see that some worked better than
others. For example scran normalisation with harmony integration seems
to produce a great overlap of batches. However, we cannot soley rely on
subjective visualisation of the batches to determine which worked best.
To help us understand the results we have included benchmarking.</p>
</div>
<div class="section level3">
<h3 id="benchmarking-integration">Benchmarking Integration<a class="anchor" aria-label="anchor" href="#benchmarking-integration"></a>
</h3>
<p>Our benchmarking involves the same methodology proposed in the SCONE.
An Average Silhouette Width assesses the closeness of the batches in
comparison to the uncorrected version.</p>
<p>previously we calcualted the UMAP reductions for the uncorrected PCA
and the corrected: harmony and scanorama data. We can use the UMAP
reductions to assess the effectiveness of the data using the
benchmarking function. Remember we can see the contents of our object
using:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/showObjectContents.html">showObjectContents</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>,<span class="st">'SCRAN'</span>,<span class="st">'SCANPY'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">SCT  contains:</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; counts</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; normalised</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; norm.scaled</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; HVGs: TRUE </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; computational_reductions: PCA </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; integration_reductions: PCA_HARMONY, SCANORAMA, CCA </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; visualisation_reductions: PCA_UMAP, PCA_HARMONY_UMAP, SCANORAMA_UMAP, CCA_UMAP, PCA_BBKNN_BBKNN:UMAP </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; neighbourhood: PCA_HARMONY_NN, SCANORAMA_NN, CCA_NN, PCA_BBKNN_BBKNN </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; cluster_assignments: PCA_HARMONY_NN:LOUVAIN, SCANORAMA_NN:LOUVAIN, CCA_NN:LOUVAIN, PCA_BBKNN_BBKNN:LOUVAIN </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; benchmarking_results: </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; (clustering) PCA_HARMONY_NN:LOUVAIN_benchmarked, SCANORAMA_NN:LOUVAIN_benchmarked, CCA_NN:LOUVAIN_benchmarked, PCA_BBKNN_BBKNN:LOUVAIN_benchmarked </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; (integration) SCT_UNCORRECTED, SCT_HARMONY, SCT_SCANORAMA, SCT_CCA, SCT_BBKNN </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt;           </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; SCRAN  contains:</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; counts</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; normalised</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; norm.scaled</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; HVGs: TRUE </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; computational_reductions: PCA </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; integration_reductions: PCA_HARMONY, SCANORAMA, CCA </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; visualisation_reductions: PCA_UMAP, PCA_HARMONY_UMAP, SCANORAMA_UMAP, CCA_UMAP, PCA_BBKNN_BBKNN:UMAP </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; neighbourhood: PCA_HARMONY_NN, SCANORAMA_NN, CCA_NN, PCA_BBKNN_BBKNN </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; cluster_assignments: PCA_HARMONY_NN:LOUVAIN, SCANORAMA_NN:LOUVAIN, CCA_NN:LOUVAIN, PCA_BBKNN_BBKNN:LOUVAIN </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; benchmarking_results: </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; (clustering) PCA_HARMONY_NN:LOUVAIN_benchmarked, SCANORAMA_NN:LOUVAIN_benchmarked, CCA_NN:LOUVAIN_benchmarked, PCA_BBKNN_BBKNN:LOUVAIN_benchmarked </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; (integration) SCRAN_UNCORRECTED, SCRAN_HARMONY, SCRAN_SCANORAMA, SCRAN_CCA, SCRAN_BBKNN </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt;           </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; SCANPY  contains:</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; counts</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; normalised</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; norm.scaled</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; HVGs: TRUE </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; computational_reductions: PCA </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; integration_reductions: PCA_HARMONY, SCANORAMA, CCA </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; visualisation_reductions: PCA_UMAP, PCA_HARMONY_UMAP, SCANORAMA_UMAP, CCA_UMAP, PCA_BBKNN_BBKNN:UMAP </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; neighbourhood: PCA_HARMONY_NN, SCANORAMA_NN, CCA_NN, PCA_BBKNN_BBKNN </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; cluster_assignments: PCA_HARMONY_NN:LOUVAIN, SCANORAMA_NN:LOUVAIN, CCA_NN:LOUVAIN, PCA_BBKNN_BBKNN:LOUVAIN </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; benchmarking_results: </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; (clustering) PCA_HARMONY_NN:LOUVAIN_benchmarked, SCANORAMA_NN:LOUVAIN_benchmarked, CCA_NN:LOUVAIN_benchmarked, PCA_BBKNN_BBKNN:LOUVAIN_benchmarked </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; (integration) SCANPY_UNCORRECTED, SCANPY_HARMONY, SCANPY_SCANORAMA, SCANPY_CCA, SCANPY_BBKNN </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt;           </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; </span></span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Here we are benchmarking how effective the correction methods were at removing batch effects. Producing the necessary distance matrices can be</span></span>
<span><span class="co"># computationally demanding, therefore we included the ability to spread the computations across different cores using the threads parameter.  </span></span>
<span></span>
<span><span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/benchmark.integration.html">benchmark.integration</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, </span>
<span>                                  batch <span class="op">=</span> <span class="st">'original.project'</span>, assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>,<span class="st">'SCRAN'</span>,<span class="st">'SCANPY'</span><span class="op">)</span>, </span>
<span>                                  reduction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'PCA_UMAP'</span>, <span class="st">'PCA_HARMONY_UMAP'</span>, </span>
<span>                                                <span class="st">'SCANORAMA_UMAP'</span>, <span class="st">'CCA_UMAP'</span>, </span>
<span>                                                <span class="st">'PCA_BBKNN_BBKNN:UMAP'</span><span class="op">)</span>, </span>
<span>                                  result.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'UNCORRECTED'</span>, <span class="st">'HARMONY'</span>, </span>
<span>                                                   <span class="st">'SCANORAMA'</span>,<span class="st">'CCA'</span>,<span class="st">'BBKNN'</span><span class="op">)</span>, </span>
<span>                                  n.components <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                                  threads <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># we can plot the results with the following command</span></span>
<span></span>
<span><span class="fu">plot.integration.benchmarking</span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>,<span class="st">'SCRAN'</span>,<span class="st">'SCANPY'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="integration_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<p>For Average Silhouette Width (ASW) when observing clustering
(biological effects) we wish to see a higher value. However, when
reducing batches we wish to see a more negative value (even a negative
value). As you can see, the uncorrected version has a high ASW value
which indicates profound batch effects are present. Here was can see
that since the result for the uncorrected data is extremely high there
are batch effects present, in comparison we have observed a reduction in
batch effects when we have used different normalisation and integration
methods. For this analysis, we have seen that Seurat CCA seems to have
aggressively removed the most batch effects whulst scanorama and BBKNN
seemed to have the least effect.</p>
<p>With these datasets, we also have ground truth labels which enbales
us to identify the best clustering labels using some benchmarking
metrices.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Here we are benchmarking how effective the correction methods were at removing batch effects. Producing the necessary distance matrices can be</span></span>
<span><span class="co"># computationally demanding, therefore we included the ability to spread the computations across different cores using the threads parameter.  </span></span>
<span></span>
<span><span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/benchmark.clustering.html">benchmark.clustering</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, </span>
<span>                                 assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>,<span class="st">'SCRAN'</span>,<span class="st">'SCANPY'</span><span class="op">)</span>, </span>
<span>                                 reduction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'PCA_HARMONY_UMAP'</span>, <span class="st">'SCANORAMA_UMAP'</span>, </span>
<span>                                               <span class="st">'CCA_UMAP'</span>, <span class="st">'PCA_BBKNN_BBKNN:UMAP'</span><span class="op">)</span>, </span>
<span>                                 clustering <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'PCA_HARMONY_NN:LOUVAIN'</span>, <span class="st">'SCANORAMA_NN:LOUVAIN'</span>, </span>
<span>                                                <span class="st">'CCA_NN:LOUVAIN'</span>, <span class="st">'PCA_BBKNN_BBKNN:LOUVAIN'</span><span class="op">)</span>, </span>
<span>                                 n.dims <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                                 ground.truth.column <span class="op">=</span> <span class="st">'celltype'</span>, </span>
<span>                                 threads <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Here we will compare pipeline performance compared to the ground truth, we will focus on ARI where the highest value is the closest to match</span></span>
<span><span class="co"># the ground truth. </span></span>
<span></span>
<span><span class="va">benchmark_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/integrate.cluster.benchmarking.results.html">integrate.cluster.benchmarking.results</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># We then order the results to haev the highest ARI scores at the top </span></span>
<span></span>
<span><span class="va">benchmarking_result</span> <span class="op">&lt;-</span> <span class="va">benchmark_results</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span>decreasing <span class="op">=</span> <span class="cn">T</span>, <span class="va">benchmark_results</span><span class="op">$</span><span class="va">ARI</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># And then we check the top 10 results</span></span>
<span></span>
<span><span class="va">benchmarking_result</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="op">]</span></span></code></pre></div>
<p>Here, we can see that SCT_PCA_BBKNN_BBKNN:LOUVAIN_benchmarked.res_0.2
performed the best compared to all other pipelines. Lets compare the
cluster assignments vs the ground truth.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">plot.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot.list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'PCA_BBKNN_BBKNN:UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCT'</span>,</span>
<span>                                   clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'celltype'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'Ground Truth'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot.list</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'PCA_BBKNN_BBKNN:UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCT'</span>,</span>
<span>                                   clust.method <span class="op">=</span> <span class="st">'PCA_BBKNN_BBKNN:LOUVAIN'</span>, column <span class="op">=</span> <span class="st">'res_0.2'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'Best Pipeline'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">egg</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/egg/man/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plots <span class="op">=</span> <span class="va">plot.list</span>, nrow <span class="op">=</span> <span class="fl">1</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="integration_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<p>However, ground truth labels are seldom available for datasets thus
the clusters must be identified manually using the feature plotting
functions and differential expression which are explained in sections 10
of the getting started tutorial. Remember to change your assays and
cluster assignment parameters!</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Connor Knight, Faraz Khan.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
