<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Data Integration â€¢ IBRAP</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Data Integration">
<meta property="og:description" content="IBRAP">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">IBRAP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/installation_linux.html">Installation (linux)</a>
    </li>
    <li>
      <a href="../articles/installation_macos.html">Installation (macos)</a>
    </li>
    <li>
      <a href="../articles/using_IBRAP.html">Getting Started</a>
    </li>
    <li>
      <a href="../articles/integration.html">Data Integration</a>
    </li>
    <li>
      <a href="../articles/application.html">RShiny Application</a>
    </li>
    <li>
      <a href="../articles/trajectory.html">Trajectory Inference</a>
    </li>
    <li>
      <a href="../articles/singleR_annotation.html">Automated Annotation</a>
    </li>
    <li>
      <a href="../articles/data_transform.html">Data Transformation</a>
    </li>
    <li>
      <a href="../articles/clustering_method.html">Clustering Methods</a>
    </li>
    <li>
      <a href="../articles/HPC_issues.html">HPC issues/solutions</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="integration_files/clipboard-1.7.1/clipboard.min.js"></script><link href="integration_files/primer-tooltips-1.4.0/build.css" rel="stylesheet">
<link href="integration_files/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet">
<script src="integration_files/klippy-0.0.0.9500/js/klippy.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Data Integration</h1>
                        <h4 data-toc-skip class="author">Connor H Knight &amp; Faraz Khan</h4>
            
            <h4 data-toc-skip class="date">11/08/2021</h4>
      
      
      <div class="hidden name"><code>integration.Rmd</code></div>

    </div>

    
    
<script>
  addClassKlippyTo("pre.r, pre.markdown");
  addKlippy('right', 'bottom', 'auto', '1', 'Click to copy', 'Done');
</script><p>It is common in single-cell analyses to integrate multiple samples. There are numerous package available for do this, however they function differently and generate different results.Therefore, it is becoming necessary to benchmark a few options before deciding on a method. For this, IBRAP contains 4 integration methods and benchmarking capabilities to distinguish the best solution.</p>
<div class="section level2">
<h2 id="integrating-multiple-datasets">Integrating Multiple Datasets<a class="anchor" aria-label="anchor" href="#integrating-multiple-datasets"></a>
</h2>
<p>For this tutorial, we have supplied two well characterised pancreas samples: CELseq2 and Smartseq2 which contain 2285 and 2394 cells, respectively.</p>
<p>You can download the data for this tutorial using the following links:</p>
<p><a href="https://www.dropbox.com/s/heughk2kl0lg6qc/smartseq2.rds?dl=0" class="external-link uri">https://www.dropbox.com/s/heughk2kl0lg6qc/smartseq2.rds?dl=0</a> <a href="https://www.dropbox.com/s/txlq7875wkthb3y/celseq2.rds?dl=0" class="external-link uri">https://www.dropbox.com/s/txlq7875wkthb3y/celseq2.rds?dl=0</a></p>
<div class="section level3">
<h3 id="creating-ibrap-objects">Creating IBRAP objects<a class="anchor" aria-label="anchor" href="#creating-ibrap-objects"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">IBRAP</span><span class="op">)</span>

<span class="va">celseq2_items</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'/Users/knight05/work/Results/scRNA-seq/IBRAP_tutorials/integration_analysis/celseq2.rds'</span><span class="op">)</span>
<span class="va">smartseq2_items</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'/Users/knight05/work/Results/scRNA-seq/IBRAP_tutorials/integration_analysis/smartseq2.rds'</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># Here we set the maximum amount of RAM R may use, for this we have allowed 4GB</span>

<span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>future.globals.maxSize <span class="op">=</span> <span class="fl">4000</span> <span class="op">*</span> <span class="fl">1024</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span>

<span class="va">celseq2_items</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'~/path/to/celseq2/celseq2.rds'</span><span class="op">)</span>
<span class="va">smartseq2_items</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'~/path/to/celseq2/celseq2.rds'</span><span class="op">)</span>

<span class="va">celseq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createIBRAPobject.html">createIBRAPobject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">celseq2</span>,
                             meta.data <span class="op">=</span> <span class="va">celseq2_items</span><span class="op">$</span><span class="va">metadata</span>,
                             original.project <span class="op">=</span> <span class="st">'celseq2'</span>,
                             method.name <span class="op">=</span> <span class="st">'RAW'</span>,
                             min.cells <span class="op">=</span> <span class="fl">3</span>,
                             min.features <span class="op">=</span> <span class="fl">200</span><span class="op">)</span>

<span class="va">smartseq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createIBRAPobject.html">createIBRAPobject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">smartseq2</span>, 
                               meta.data <span class="op">=</span> <span class="va">smartseq2_items</span><span class="op">$</span><span class="va">metadata</span>,
                               original.project <span class="op">=</span> <span class="st">'smartseq2'</span>,
                               method.name <span class="op">=</span> <span class="st">'RAW'</span>,
                               min.cells <span class="op">=</span> <span class="fl">3</span>,
                               min.features <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="merging-ibrap-objects">Merging IBRAP objects<a class="anchor" aria-label="anchor" href="#merging-ibrap-objects"></a>
</h3>
<p>Datasets are then merged together. WARNING: Datasets can only be merged straight after their creation with no downstream results.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/merge.html">merge</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">celseq2</span>, y <span class="op">=</span> <span class="va">smartseq2</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="pre-processing-batch-visualisation">Pre-processing &amp; Batch Visualisation<a class="anchor" aria-label="anchor" href="#pre-processing-batch-visualisation"></a>
</h3>
<p>We then continue our analysis as normal up until PCA reduction:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.sct.html">perform.sct</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, 
                        assay <span class="op">=</span> <span class="st">'RAW'</span>, 
                        slot <span class="op">=</span> <span class="st">'counts'</span><span class="op">)</span>

<span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.scran.html">perform.scran</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, 
                          assay <span class="op">=</span> <span class="st">'RAW'</span>, 
                          slot <span class="op">=</span> <span class="st">'counts'</span>, 
                          vars.to.regress <span class="op">=</span> <span class="st">'RAW_total.counts'</span><span class="op">)</span>

<span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.scanpy.html">perform.scanpy</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, 
                           vars.to.regress <span class="op">=</span> <span class="st">'RAW_total.counts'</span><span class="op">)</span>

<span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.pca.html">perform.pca</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, 
                        assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, 
                        n.pcs <span class="op">=</span> <span class="fl">50</span>, reduction.save <span class="op">=</span> <span class="st">'pca'</span><span class="op">)</span>

<span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.umap.html">perform.umap</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, 
                         assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, 
                         reduction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'pca'</span><span class="op">)</span>, 
                         n_components <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<p>Now that we have pre-processed our data, lets take a look at the unintegrated batches:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'PCA_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCT'</span>,
                          clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCT'</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span>


<span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'PCA_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCRAN'</span>,
                          clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCRAN'</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span>


<span class="va">plot3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'PCA_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCANPY'</span>,
                          clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCANPY'</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span>

<span class="fu">egg</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/egg/man/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plots <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">plot1</span>, <span class="va">plot2</span>, <span class="va">plot3</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>
                        </code></pre></div>
<p><img src="unintegrated.png" width="150%" height="150%"></p>
<p>As we can see there are profound differences between the samples due to technical variation. This requires correction.</p>
</div>
<div class="section level3">
<h3 id="integration-techniques">Integration Techniques<a class="anchor" aria-label="anchor" href="#integration-techniques"></a>
</h3>
<p>We have included 4 integration techniques:</p>
<ul>
<li>Seurat CCA (counts)</li>
<li>Scanorama (counts)</li>
<li>BBKNN (reduction)</li>
<li>Harmony (reduction)</li>
</ul>
<p>Some integration methods either correct the normalised counts or the reduction components. For this tutorial, we will compare Scanorama which corrects the counts and Harmony which corrects reductions.</p>
</div>
<div class="section level3">
<h3 id="integration-application-visualisaiton">Integration Application &amp; Visualisaiton<a class="anchor" aria-label="anchor" href="#integration-application-visualisaiton"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.harmony.html">perform.harmony</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, 
                            assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCRAN'</span>, <span class="st">'SCT'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, 
                            vars.use <span class="op">=</span> <span class="st">'original.project'</span>, 
                            reduction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'pca'</span><span class="op">)</span>,  
                            max.iter.harmony <span class="op">=</span> <span class="fl">100</span>,
                            dims.use <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="cn">NULL</span><span class="op">)</span><span class="op">)</span>
                            </code></pre></div>
<p>Harmony accepts reduced embeddings and produces a new reduced embedding called harmony reducitons. These reductions can then be used to produce nearest neighbourhood graphs, any clusxtering method, and any reduction method. Harmony functions well since you can adjust the aggressiveness of the correction per covariate that you specify using the theta parameter.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.scanorama.html">perform.scanorama</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, 
                              assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, 
                              slot <span class="op">=</span> <span class="st">'norm.scaled'</span>, 
                              split.by <span class="op">=</span> <span class="st">'original.project'</span>, 
                              n.dims <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></code></pre></div>
<p>Rather than functioning on reduced embeddings, scanorama uses the <span class="citation">@norm.scaled</span> slot of the method assay and stitches together the batches in order of similarity. Scanorama produces a reduced embedding that can be used for nearest neighbourhood graph generation, any clustering method, and non-linear reduction method.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R">                              
<span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.umap.html">perform.umap</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, 
                         assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, 
                         reduction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'pca_harmony'</span>, <span class="st">'scanorama'</span><span class="op">)</span>, 
                         n_components <span class="op">=</span> <span class="fl">2</span>, 
                         n.dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">50</span>, <span class="fl">50</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>We next created UMAPP reduced embeddings from the previously generated corrected reduced embeddings, including: PCA_HARMONY and SCANORAMA.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">plot.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">plot.list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'PCA_HARMONY_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCT'</span>,
                                   clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCT_harmony'</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span>

<span class="va">plot.list</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'PCA_HARMONY_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCRAN'</span>,
                                   clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCRAN_harmony'</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span>

<span class="va">plot.list</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'PCA_HARMONY_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCANPY'</span>,
                                   clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCANPY_harmony'</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span>

<span class="va">plot.list</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'SCANORAMA_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCT'</span>,
                                   clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCT_SCANORAMA'</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span>

<span class="va">plot.list</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'SCANORAMA_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCRAN'</span>,
                                   clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCRAN_SCANORAMA'</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span>

<span class="va">plot.list</span><span class="op">[[</span><span class="fl">6</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, reduction <span class="op">=</span> <span class="st">'SCANORAMA_UMAP'</span>, assay <span class="op">=</span> <span class="st">'SCANPY'</span>,
                                   clust.method <span class="op">=</span> <span class="st">'metadata'</span>, column <span class="op">=</span> <span class="st">'original.project'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCANPY_SCANORAMA'</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span>

<span class="fu">egg</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/egg/man/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plots <span class="op">=</span> <span class="va">plot.list</span>, nrow <span class="op">=</span> <span class="fl">2</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<p><img src="integration_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p><img src="integration.png" width="150%" height="150%"></p>
<p>From initial inspection we can see that some worked better than others. For example scran normalisation with harmony integration seems to produce a great overlap of batches. However, we cannot soley rely on subjective visualisation of the batches to determine which worked best. To help us understand the results we have included benchmarking.</p>
</div>
<div class="section level3">
<h3 id="benchmarking-integration">Benchmarking Integration<a class="anchor" aria-label="anchor" href="#benchmarking-integration"></a>
</h3>
<p>Our benchmarking involves the same methodology proposed in the SCONE. An Average Silhouette Width assesses the closeness of the batches in comparison to the uncorrected version.</p>
<p>previously we calcualted the UMAP reductions for the uncorrected PCA and the corrected: harmony and scanorama data. We can use the UMAP reductions to assess the effectiveness of the data using this function:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">pancreas</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/benchmark.integration.html">benchmark.intergation</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, 
                                  batch <span class="op">=</span> <span class="st">'original.project'</span>, assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>,<span class="st">'SCRAN'</span>,<span class="st">'SCANPY'</span><span class="op">)</span>, 
                                  reduction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'PCA_UMAP'</span>, <span class="st">'PCA_HARMONY_UMAP'</span>, 
                                                <span class="st">'SCANORAMA_UMAP'</span><span class="op">)</span>, 
                                  result.names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'uncorrected'</span>, <span class="st">'harmony'</span>, <span class="st">'scanorama'</span><span class="op">)</span>, 
                                  n.components <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>

<span class="co"># we can plot the results with the following command</span>

<span class="fu">plot.integration.benchmarking</span><span class="op">(</span>object <span class="op">=</span> <span class="va">pancreas</span>, assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>,<span class="st">'SCRAN'</span>,<span class="st">'SCANPY'</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="integration_benchmarking.png" width="150%" height="150%"></p>
<p>For Average Silhouette Width (ASW) when observing clustering (biological effects) we wish to see a higher value. However, when reducing batches we wish to see a more negative value (even a negative value). As you can see, the uncorrected version has a high ASW value which indicates profound batch effects are present. Whereas, we can see the batch correction methods have reducd the batch effects to varying degrees. For example, harmony seems to reduced the batch effect the best across all normalisation methods. Scanorama however performed less successfully. However, you should not use this as an absolute recommendation for which reduction to proceed with.</p>
<p>From this point, we can continue the analysis as usual, using these integrated reductions to cluster our cells and identify their biological assignments.</p>
<p>Benchmarking integration is useful since you can evidently witness that integration methods behave differently and produce varying results. IBRAP enables you to benchmark four different techniques and select the best option for your dataset!</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Connor Knight, Faraz Khan.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
