<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getting Started â€¢ IBRAP</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Getting Started">
<meta property="og:description" content="IBRAP">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">IBRAP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Installation_linux.html">Installation (linux)</a>
    </li>
    <li>
      <a href="../articles/installation_macos.html">Installation (macos)</a>
    </li>
    <li>
      <a href="../articles/dockers.html">Using Docker/Singularity</a>
    </li>
    <li>
      <a href="../articles/using_IBRAP.html">Getting Started</a>
    </li>
    <li>
      <a href="../articles/integration.html">Data Integration</a>
    </li>
    <li>
      <a href="../articles/application.html">RShiny Application</a>
    </li>
    <li>
      <a href="../articles/trajectory.html">Trajectory Inference</a>
    </li>
    <li>
      <a href="../articles/singleR_annotation.html">Automated Annotation</a>
    </li>
    <li>
      <a href="../articles/data_transform.html">Data Transformation</a>
    </li>
    <li>
      <a href="../articles/clustering_method.html">Clustering Methods</a>
    </li>
    <li>
      <a href="../articles/HPC_issues.html">HPC issues/solutions</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="using_IBRAP_files/clipboard-1.7.1/clipboard.min.js"></script><link href="using_IBRAP_files/primer-tooltips-1.4.0/build.css" rel="stylesheet">
<link href="using_IBRAP_files/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet">
<script src="using_IBRAP_files/klippy-0.0.0.9500/js/klippy.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Getting Started</h1>
                        <h4 data-toc-skip class="author">Connor H Knight
&amp; Faraz Khan</h4>
            
            <h4 data-toc-skip class="date">11/08/2021</h4>
      
      
      <div class="hidden name"><code>using_IBRAP.Rmd</code></div>

    </div>

    
    
<script>
  addClassKlippyTo("pre.r, pre.markdown");
  addKlippy('right', 'bottom', 'auto', '1', 'Click to copy', 'Done');
</script><p>Welcome to IBRAPs Getting Started tutorial page that teaches you how
to use some fundamental functions in IBRAP. IBRAP combines analytical
components from different packages in R and python to create new
pipelines that analyse their datasets better! Please enjoy our tutorial
and if you require any help please leave a comment on our GitHub.</p>
<div class="section level2">
<h2 id="analysing-2994-bone-marrow-mononuclear-cells-bmmc-from-a-single-sample-">Analysing 2994 Bone Marrow Mononuclear Cells (BMMC) from a single
sample.<a class="anchor" aria-label="anchor" href="#analysing-2994-bone-marrow-mononuclear-cells-bmmc-from-a-single-sample-"></a>
</h2>
<p>To demonstrate IBRAP we enlisted a sample produced by <a href="https://insight.jci.org/articles/view/124928" class="external-link">Oetjen</a> <em>et
al</em>. This can be downloaded using this <a href="https://github.com/connorhknight/IBRAP/raw/main/data/marrow_A.rds" class="external-link">link</a></p>
<p>Marrow_A is a bone marrow sample from a set of 25. It contains 2994
cells, sequenced using 10x Genomics and aligned using CellRanger. Please
download the sample and load it into R using the following command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">IBRAP</span><span class="op">)</span></span>
<span></span>
<span><span class="va">marrow_A</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'~/path/to/marrow_A.rds'</span><span class="op">)</span></span>
<span>  </span></code></pre></div>
<div class="section level3">
<h3 id="section-1-cleaning-droplet-based-datasets">Section 1: Cleaning droplet-based datasets<a class="anchor" aria-label="anchor" href="#section-1-cleaning-droplet-based-datasets"></a>
</h3>
<div class="section level4">
<h4 id="section-1-1-applying-the-functions">Section 1.1: Applying the functions<a class="anchor" aria-label="anchor" href="#section-1-1-applying-the-functions"></a>
</h4>
<p>Droplet-based datasets present unique confounding effects, including:
doublets and ambient RNA contamination. To remove them we have enlisted
scrublet and decontX, respectively. Thus far, there is no consensus on
the order of application. However, we prefer scrublet followed by
decontX.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># We simply supply our generated count matrix to both functions</span></span>
<span></span>
<span><span class="va">marrow_A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.scrublet.html">perform.scrublet</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">marrow_A</span>, print.plot <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p>Scrublet functions by simulating multiplets from out datasets (by
integrating similar cells); we can see these simulated cells with high
doublet scores. For this example, we can see a minimal number of cells
in our observed cells with high doublet scores. Cells that are above a
certain doublet score threshold are omitted from the dataset.</p>
<p>Now that we have removed these confounding effects from our dataset
we can continue with our analysis.</p>
</div>
</div>
<div class="section level3">
<h3 id="section-2-initiating-the-ibrap-s4-object">Section 2: Initiating the IBRAP S4 object<a class="anchor" aria-label="anchor" href="#section-2-initiating-the-ibrap-s4-object"></a>
</h3>
<p>Lets start by initiating our IBRAP class object ready for downstream
analyses. You can see the structure and a manual for these in our
reference section.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># whilst creating the object we begin filtering the cells and genes using min.cells and min.features</span></span>
<span><span class="va">marrow_A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createIBRAPobject.html">createIBRAPobject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">marrow_A</span>,</span>
<span>                              original.project <span class="op">=</span> <span class="st">'bmmc'</span>,</span>
<span>                              min.cells <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                              min.features <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span>
<span>                          </span></code></pre></div>
<p>This produces an S4 class object that initially contains sample
metadata and a raw method assay contained in <a href="mailto:object@methods" class="email">object@methods</a>. A couple
of important steps happen here: metadata, including: total counts and
total genes per cell are stored in <a href="mailto:object@sample" class="email">object@sample</a>_metadata, and some initial filtering of
the count matrix happens. In this case, we removed any genes that are
not present in at least 3 cells and removed cells with less than 200
genes, We performed the latter as to remove any droplets that are empty
and do not contain true cells.</p>
</div>
<div class="section level3">
<h3 id="section-3-quality-control">Section 3: Quality Control<a class="anchor" aria-label="anchor" href="#section-3-quality-control"></a>
</h3>
<div class="section level4">
<h4 id="section-3-1-metrics">Section 3.1: Metrics<a class="anchor" aria-label="anchor" href="#section-3-1-metrics"></a>
</h4>
<p>Next, we must assess the quality of our datasets, if you wish to read
in-depth about Quality Control (QC) then please refer to this <a href="https://www.embopress.org/doi/full/10.15252/msb.20188746" class="external-link">publication</a>.
In summary, we use the following metrices to QC our data:</p>
<table class="table">
<colgroup>
<col width="12%">
<col width="87%">
</colgroup>
<thead><tr class="header">
<th align="left">QC.Metric</th>
<th align="left">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Mitochondrial Fraction</td>
<td align="left">An abnormally high mitochondrial fraction indicates
improper wet lab preparation</td>
</tr>
<tr class="even">
<td align="left">Total RNA</td>
<td align="left">A difference in count value can affect downstream
processes and requires normalising downstream</td>
</tr>
<tr class="odd">
<td align="left">Total Genes</td>
<td align="left">Low gene presence indicates droplets containing no
cells whereas, disproportionately high number of genes indicates a
droplet containing more than one cell</td>
</tr>
</tbody>
</table>
<p>As previously mentioned we already generated total count and total
genes whilst creating our IBRAP object. However, percentage
mitochondrial genes requires a second function since mitochondrial genes
have a varying prefix in datasets:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">marrow_A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_percentage_genes.html">find_percentage_genes</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">marrow_A</span>, pattern <span class="op">=</span> <span class="st">'^MT-'</span>,</span>
<span>                                  assay <span class="op">=</span> <span class="st">'RAW'</span>, </span>
<span>                                  slot <span class="op">=</span> <span class="st">'counts'</span>,</span>
<span>                                  column.name <span class="op">=</span> <span class="st">'RAW_percent.mt'</span><span class="op">)</span></span>
<span>                              </span></code></pre></div>
</div>
<div class="section level4">
<h4 id="section-3-2-visualisation">Section 3.2: Visualisation<a class="anchor" aria-label="anchor" href="#section-3-2-visualisation"></a>
</h4>
<p>Visualising this information is important for high dimensional data.
Thus, we can achieve this in two ways: using violin plots or comparing
two metrics with scatter plots.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/plot.QC.vln.html">plot.QC.vln</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">marrow_A</span>, </span>
<span>            metadata.columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'RAW_total.features'</span>, </span>
<span>                                 <span class="st">'RAW_total.counts'</span>, </span>
<span>                                 <span class="st">'RAW_percent.mt'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="using_IBRAP_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>The violin plot is split into its metrics and displays the volume of
cells expressing the value on the y-axis.</p>
<p>Scatter plots enable us to observe correlations between metrics,
i.e.Â the higher the amount of features in a cell the higher the number
of counts there should be.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/plot.QC.scatter.html">plot.QC.scatter</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">marrow_A</span>, </span>
<span>                x <span class="op">=</span> <span class="st">'RAW_total.counts'</span>, </span>
<span>                y <span class="op">=</span> <span class="st">'RAW_total.features'</span>, </span>
<span>                split.by <span class="op">=</span> <span class="st">'original.project'</span></span>
<span>                <span class="op">)</span></span></code></pre></div>
<p><img src="using_IBRAP_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>Now that we have eye-balled our data we can proceed to
filtration.</p>
</div>
<div class="section level4">
<h4 id="section-3-3-filtration">Section 3.3: Filtration<a class="anchor" aria-label="anchor" href="#section-3-3-filtration"></a>
</h4>
<p>We clean our data further by applyng filtration cut-offs. We remove
cells with a high number of features since this will remove any doublets
that were missed by scrublet (two cells in a droplet will display a
disproportionately higher number of features). Then, we remove cells
with a certain fraction of mitochondrial counts - we do this to remove
cells with damaged cell membrane. The percentage cut-off for
mitochondrial genes is set at a case-by-case basis and no single value
is applicable to all samples.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">marrow_A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_IBRAP.html">filter_IBRAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">marrow_A</span>, <span class="va">RAW_total.features</span> <span class="op">&lt;</span> <span class="fl">3000</span> <span class="op">&amp;</span> <span class="va">RAW_percent.mt</span> <span class="op">&lt;</span> <span class="fl">8</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="section-4-data-transformation">Section 4: Data Transformation<a class="anchor" aria-label="anchor" href="#section-4-data-transformation"></a>
</h3>
<div class="section level4">
<h4 id="section-4-1-applying-the-functions">Section 4.1: Applying The Functions<a class="anchor" aria-label="anchor" href="#section-4-1-applying-the-functions"></a>
</h4>
<p>To accommodate downstream analyses, we must promote a normal
distribution and numerical scaling in our data. We included four
separate <a href="https://connorhknight.github.io/IBRAP/articles/data_transform.html" class="external-link">normalisation</a>
methods that are applied as such:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">marrow_A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.sct.html">perform.sct</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">marrow_A</span>, </span>
<span>                        vars.to.regress <span class="op">=</span> <span class="st">'RAW_percent.mt'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">marrow_A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.scran.html">perform.scran</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">marrow_A</span>, </span>
<span>                          vars.to.regress <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'RAW_percent.mt'</span>, <span class="st">'RAW_total.counts'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">marrow_A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.scanpy.html">perform.scanpy</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">marrow_A</span>, </span>
<span>                           vars.to.regress <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'RAW_percent.mt'</span>, <span class="st">'RAW_total.counts'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Now, we can see the available normalisation assays through the
following command:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">marrow_A</span><span class="op">@</span><span class="va">sample_metadata</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="op">]</span></span>
<span><span class="co">#&gt;                    original.project RAW_total.counts RAW_total.features</span></span>
<span><span class="co">#&gt; AAACCTGCAAGTAATG-1             bmmc            10568                519</span></span>
<span><span class="co">#&gt; AAACCTGCAGCGAACA-1             bmmc             6378               1861</span></span>
<span><span class="co">#&gt; AAACCTGCATGTCCTC-1             bmmc             4172               1082</span></span>
<span><span class="co">#&gt; AAACCTGGTCGACTGC-1             bmmc             6608               1618</span></span>
<span><span class="co">#&gt; AAACCTGGTCGCTTCT-1             bmmc             5034               1413</span></span>
<span><span class="co">#&gt; AAACCTGTCCCGACTT-1             bmmc             3998               1127</span></span>
<span><span class="co">#&gt; AAACGGGAGGGCACTA-1             bmmc             4401               1185</span></span>
<span><span class="co">#&gt; AAACGGGCACCTCGGA-1             bmmc             9226               2011</span></span>
<span><span class="co">#&gt; AAACGGGCACGTGAGA-1             bmmc             4352               1231</span></span>
<span><span class="co">#&gt; AAACGGGCATACGCTA-1             bmmc             4200                998</span></span>
<span><span class="co">#&gt;                    RAW_percent.mt SCT_total.counts SCT_total.features</span></span>
<span><span class="co">#&gt; AAACCTGCAAGTAATG-1      0.1230129         485.3213                518</span></span>
<span><span class="co">#&gt; AAACCTGCAGCGAACA-1      4.7977422        1957.0735               1859</span></span>
<span><span class="co">#&gt; AAACCTGCATGTCCTC-1      2.4928092        1138.8269               1082</span></span>
<span><span class="co">#&gt; AAACCTGGTCGACTGC-1      5.1906780        1732.1269               1616</span></span>
<span><span class="co">#&gt; AAACCTGGTCGCTTCT-1      4.1716329        1498.7047               1411</span></span>
<span><span class="co">#&gt; AAACCTGTCCCGACTT-1      4.3521761        1181.2918               1126</span></span>
<span><span class="co">#&gt; AAACGGGAGGGCACTA-1      4.0672574        1282.3847               1185</span></span>
<span><span class="co">#&gt; AAACGGGCACCTCGGA-1      3.2083243        2011.5219               2006</span></span>
<span><span class="co">#&gt; AAACGGGCACGTGAGA-1      4.4117647        1303.8643               1230</span></span>
<span><span class="co">#&gt; AAACGGGCATACGCTA-1      2.9047619        1077.4954                996</span></span>
<span><span class="co">#&gt;                    SCRAN_total.counts SCRAN_total.features SCANPY_total.counts</span></span>
<span><span class="co">#&gt; AAACCTGCAAGTAATG-1           776.7584                  519            521.5231</span></span>
<span><span class="co">#&gt; AAACCTGCAGCGAACA-1          2836.5209                 1861           2508.5597</span></span>
<span><span class="co">#&gt; AAACCTGCATGTCCTC-1          1612.1623                 1082           1755.9816</span></span>
<span><span class="co">#&gt; AAACCTGGTCGACTGC-1          2517.0118                 1618           2174.6400</span></span>
<span><span class="co">#&gt; AAACCTGGTCGCTTCT-1          2156.1500                 1413           2143.2267</span></span>
<span><span class="co">#&gt; AAACCTGTCCCGACTT-1          1667.6693                 1127           1858.7245</span></span>
<span><span class="co">#&gt; AAACGGGAGGGCACTA-1          1822.0787                 1185           1925.6957</span></span>
<span><span class="co">#&gt; AAACGGGCACCTCGGA-1          3024.1470                 2011           2192.6501</span></span>
<span><span class="co">#&gt; AAACGGGCACGTGAGA-1          1854.9666                 1231           1979.5510</span></span>
<span><span class="co">#&gt; AAACGGGCATACGCTA-1          1528.2787                  998           1646.3604</span></span>
<span><span class="co">#&gt;                    SCANPY_total.features</span></span>
<span><span class="co">#&gt; AAACCTGCAAGTAATG-1                   519</span></span>
<span><span class="co">#&gt; AAACCTGCAGCGAACA-1                  1861</span></span>
<span><span class="co">#&gt; AAACCTGCATGTCCTC-1                  1082</span></span>
<span><span class="co">#&gt; AAACCTGGTCGACTGC-1                  1618</span></span>
<span><span class="co">#&gt; AAACCTGGTCGCTTCT-1                  1413</span></span>
<span><span class="co">#&gt; AAACCTGTCCCGACTT-1                  1127</span></span>
<span><span class="co">#&gt; AAACGGGAGGGCACTA-1                  1185</span></span>
<span><span class="co">#&gt; AAACGGGCACCTCGGA-1                  2011</span></span>
<span><span class="co">#&gt; AAACGGGCACGTGAGA-1                  1231</span></span>
<span><span class="co">#&gt; AAACGGGCATACGCTA-1                   998</span></span></code></pre></div>
<p>Please note, that if your computer has small RAM, then you can change
the conserve.memory parameter in sctransform to TRUE, this will make the
function slower by more memory efficient.</p>
<p>For these methods, we transformed our raw counts, and regressed some
QC metrics from our datasets to not allow them to confound our
downstream analyses. There are some notable differences here. For
example, SCT automatically regresses RAW_total.counts so we do not need
to specify this however, the others you do. These parameters are set as
default, any deviation from these is at your own risk since they could
produce incorrect results. Nonetheless, feel free to play around and
observe the difference.</p>
</div>
<div class="section level4">
<h4 id="section-4-2-how-functions-work-downstream">Section 4.2: How Functions Work Downstream<a class="anchor" aria-label="anchor" href="#section-4-2-how-functions-work-downstream"></a>
</h4>
<p>Each normalisation function produces a new method-assay which
contains the results. Any downstream analyses will be stored in these
separate compartments, respective to the normalisation method. i.e.Â <a href="mailto:object@methods" class="email">object@methods</a> contains
all of the method assays. Hereon, function will iterate through the
indicated method-assays and any other inputs within the assays. For
example, if you supply c(â€˜SCTâ€™,â€˜SCRANâ€™,â€˜SCANPYâ€™) as assays and
c(â€˜pca_umapâ€™, â€˜pca_t-SNEâ€™) as reductions. A function will perform
firstly on â€˜pca_umapâ€™ in the â€˜SCTâ€™ assay, then the â€˜pca_t-SNEâ€™ in the
â€˜SCTâ€™ assay. After these the function will move on to â€˜pca_umapâ€™ in the
â€˜SCRANâ€™ assay and so on. You will see this as we progress through the
tutorial.</p>
</div>
<div class="section level4">
<h4 id="section-4-3-downstream-pipeline-deviations">Section 4.3: Downstream Pipeline Deviations<a class="anchor" aria-label="anchor" href="#section-4-3-downstream-pipeline-deviations"></a>
</h4>
<p>Since IBRAP incorporated multiple pipelines, some pipelines begin to
vary at this point. For example, Scanpy PCA calculation and
neighbourhood graph generation slightly differ. IBRAP accomodates this
by identifyig the assay name and automatically applying the correct
function. When adding suffixes to any result naming it is important not
to include your own underscores.</p>
</div>
</div>
<div class="section level3">
<h3 id="section-5-linear-dimensionaly-reduction">Section 5: Linear Dimensionaly Reduction<a class="anchor" aria-label="anchor" href="#section-5-linear-dimensionaly-reduction"></a>
</h3>
<div class="section level4">
<h4 id="section-5-1-principal-component-analysis">Section 5.1: Principal Component Analysis<a class="anchor" aria-label="anchor" href="#section-5-1-principal-component-analysis"></a>
</h4>
<p>Now that our data has been appropriately transformed our data and
highly variable genes have been identified we can reduce the number of
dimensions using Principal Component Analysis (PCA). Reduction methods
summarise as much information from the higher dimensions into lower
dimensions as possible. Therefore, these lower dimensions facilitate a
faster computing time and require less computational resources.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">marrow_A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.pca.html">perform.pca</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">marrow_A</span>, </span>
<span>                        assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, </span>
<span>                        n.pcs <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<p>As you can see, we have requested this function to iterate through
all three assays and to save the reduction as pca. The produced
principal components (PCs) contain a different explained variance, high
variance PCs contain the most information between points and low provide
the opposite. Furthermore, low variation PCs create a false closeness
between cell points and thus require exclusion from <em>some</em>
downstream analyses. A simple way to identify the number of PCs to
retain is to plot them in descending order and find the elbow point. The
elbow point being the bend in the plot of points. We can observe the
amount of variance by plotting an elbow plot using the following
function:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/plot.variance.html">plot.variance</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">marrow_A</span>, </span>
<span>              assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>,<span class="st">'SCRAN'</span>,<span class="st">'SCANPY'</span><span class="op">)</span>, </span>
<span>              reduction <span class="op">=</span> <span class="st">'PCA'</span><span class="op">)</span></span></code></pre></div>
<p><img src="using_IBRAP_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>For these assays, the elbow point is around 10-15 PCs.</p>
</div>
</div>
<div class="section level3">
<h3 id="section-6-generating-neighbourhood-graphs-clustering">Section 6: Generating Neighbourhood Graphs &amp; Clustering<a class="anchor" aria-label="anchor" href="#section-6-generating-neighbourhood-graphs-clustering"></a>
</h3>
<div class="section level4">
<h4 id="section-6-1-neighbourhood-graphs">Section 6.1: Neighbourhood Graphs<a class="anchor" aria-label="anchor" href="#section-6-1-neighbourhood-graphs"></a>
</h4>
<p>Prior to clustering it is important to ascertain the relationship and
closeness of our cells. As a pre-requisite to clustering we are going to
produce a k-nearest neghbour graphs. SCT, SCRAN, and SCANPY use one
algorithm whilst SCANPY uses its own version. This is produced by
applying the following function:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># here we generate the neighbour graph for PCA and produce a diffusion map from the PCA-neighbour graph</span></span>
<span></span>
<span><span class="va">marrow_A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.nn.html">perform.nn</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">marrow_A</span>, </span>
<span>                       assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, </span>
<span>                       reduction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'PCA'</span><span class="op">)</span>, dims.use <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span></span>
<span>                          </span></code></pre></div>
<p>In this function, we have requested that the function iterate through
all three assays, within each assay to use pca and then pca:diffmap
using all dimensions (seen as list(0,0)). These results will be stored
under <a href="mailto:assay@neighbours" class="email">assay@neighbours</a> ready for use downstream. ### Section
6.2: Clustering</p>
<p>Finally, using our neighbourhood graph we must cluster close networks
together to infer individual cell types. perform.graph.cluster()
provides either: - Louvain algorithm (1) - Louvain algorithm with
multilevel refinement (2) - Smart Local Moving (SLM) (3) - Leiden
(4)</p>
<p>These methods can be adjusted using thre res parameter in the
function typically ranging between 0.1-1.5. A higher resolution finds
more communities in a sample, whereas a lower value produces the
opposite. The default option for this function is to use the following
resolutions: c(0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1, 1.1, 1.2,
1.3, 1.4, 1.5). However, feel free to adjsut this to your
requirement.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">marrow_A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.graph.cluster.html">perform.graph.cluster</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">marrow_A</span>, </span>
<span>                                  assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, </span>
<span>                                  neighbours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'PCA_NN'</span><span class="op">)</span>, </span>
<span>                                  algorithm <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>For this example we clustered all assays and used our neighbourhood
graphs generated from PCA and Diffusion maps using the classic louvain
algorithm. As mentioned above there are a few graph-based methods. But,
we have also inclusing a couple of <a href="https://connorhknight.github.io/IBRAP/articles/clustering_method.html" class="external-link">alternatives</a>
that are described in another tutorial</p>
</div>
</div>
<div class="section level3">
<h3 id="section-7-non-linear-reduction">Section 7: Non-linear Reduction<a class="anchor" aria-label="anchor" href="#section-7-non-linear-reduction"></a>
</h3>
<p>Now we have reduced the dimensionalty of our datasets using
non-linear methods and clustered the cells according to their similarity
we must visualise these results. However, linear methods such as PCA and
diffusion maps are inherently innapropriate for visualisation.
Therefore, we must process then using non-linear methods.</p>
<p>IBRAP integrated the ability to perform 3 methods: - UMAP - t-SNE -
lvish (a LargeVis adaptation)</p>
<p>For this tutorial we will focus on UMAP.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">marrow_A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.umap.html">perform.umap</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">marrow_A</span>, </span>
<span>                         assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, </span>
<span>                         reduction <span class="op">=</span> <span class="st">'PCA'</span>, </span>
<span>                         n.dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Simply, we asked this function to iterate through all assays, and use
the first 15 components of the pca reduction.</p>
<p>Now, we have generated a high volume of results, especially
clustering results. Therefore, it is beneficial to applying to some
benchmarking comparisons.</p>
</div>
<div class="section level3">
<h3 id="section-8-benchmarking-results">Section 8: Benchmarking Results<a class="anchor" aria-label="anchor" href="#section-8-benchmarking-results"></a>
</h3>
<p>IBRAP contains a benchmarking function that enlists multiple metrices
to assess clustering performance. These metrices are only informative
however and will be unable to distinguish the best results (that is down
to the user to decide).</p>
<table class="table">
<colgroup>
<col width="16%">
<col width="83%">
</colgroup>
<thead><tr class="header">
<th align="left">QC.Metric</th>
<th align="left">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Average Silhouette Width (ASW)</td>
<td align="left">This measures how close cells are to their own cluster
(cohesion) whilst comparing to other clusters (separation) A higher
value is good.</td>
</tr>
<tr class="even">
<td align="left">Dunn Index</td>
<td align="left">Comapres the ratio of the smallest distance between
observations not in the same cluster to the largest intra-cluster
distance. A higher value is preferred.</td>
</tr>
<tr class="odd">
<td align="left">Connectivity</td>
<td align="left">This uses a k-nearest neighbour algorithm that
determines the connectedness between clusters. A depleted value is
prefered.</td>
</tr>
</tbody>
</table>
<p>Now, lets take a look at the data that we have produced so far:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/showObjectContents.html">showObjectContents</a></span><span class="op">(</span><span class="va">marrow_A</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>,<span class="st">'SCRAN'</span>,<span class="st">'SCANPY'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">SCT  contains:</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; counts</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; normalised</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; norm.scaled</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; HVGs: TRUE </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; computational_reductions: PCA </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; integration_reductions:  </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; visualisation_reductions: PCA_UMAP </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; neighbourhood: PCA_NN </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; cluster_assignments: PCA_NN:LOUVAIN </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; benchmarking_results: </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; (clustering) PCA_NN:LOUVAIN_benchmarked </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt;           </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; SCRAN  contains:</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; counts</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; normalised</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; norm.scaled</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; HVGs: TRUE </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; computational_reductions: PCA </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; integration_reductions:  </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; visualisation_reductions: PCA_UMAP </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; neighbourhood: PCA_NN </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; cluster_assignments: PCA_NN:LOUVAIN </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; benchmarking_results: </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; (clustering) PCA_NN:LOUVAIN_benchmarked </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt;           </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; SCANPY  contains:</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; counts</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; normalised</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; norm.scaled</span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; HVGs: TRUE </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; computational_reductions: PCA </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; integration_reductions:  </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; visualisation_reductions: PCA_UMAP </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; neighbourhood: PCA_NN </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; cluster_assignments: PCA_NN:LOUVAIN </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; benchmarking_results: </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; (clustering) PCA_NN:LOUVAIN_benchmarked </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt;           </span></span></span>
<span><span class="co"><span style="color: #00BBBB;">#&gt; </span></span></span></code></pre></div>
<p>To set up the function, supply the names you want to benchmark from
cluster_assignments. Next, include the corresponding reductions for
them, normally this will be a umap reduction using 1:2 dimensions. For
example, if clustering and the umap were calculated using PCA these
should be paired, likewise if they were calculated using an integration
method we would do the same.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">marrow_A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/benchmark.clustering.html">benchmark.clustering</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">marrow_A</span>, </span>
<span>                                 assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, </span>
<span>                                 clustering <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'PCA_NN:LOUVAIN'</span><span class="op">)</span>, </span>
<span>                                 reduction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'PCA_UMAP'</span><span class="op">)</span>, </span>
<span>                                 n.dims <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p>We can visualise the comparison within a clustering method dataframe
using the following function:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">plot.cluster.benchmarking</span><span class="op">(</span>object <span class="op">=</span> <span class="va">marrow_A</span>, assay <span class="op">=</span> <span class="st">'SCT'</span>, </span>
<span>                          clustering <span class="op">=</span> <span class="st">'PCA_NN:LOUVAIN_benchmarked'</span><span class="op">)</span></span></code></pre></div>
<p><img src="using_IBRAP_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<p>This provides a line graph showing the metrices. As you can see,
there is a correlation between the results. However, do not be fooled by
the preferred valuei n very low settings, this is likely due to a
smaller number of communities being detected at lower resolutions which
means that clusters are typically more separated.</p>
<p>Now that we have our results and benchmarking to guide us we can
begin to visualise the results.</p>
</div>
<div class="section level3">
<h3 id="section-9-visualising-clustering-results">Section 9: Visualising Clustering Results<a class="anchor" aria-label="anchor" href="#section-9-visualising-clustering-results"></a>
</h3>
<p>Since we have 3 different normalisation method it is beneficial to
observe the difference in their produces topology and clustering.</p>
<p>The main challenge with the plotting function in IBRAP is isolating
the correct labelling. In general, plotting functions will ask for the
assay which is contained within <a href="mailto:object@methods" class="email">object@methods</a>, the clust.method that is contained
withing <span class="citation">@cluster_assignments</span>, and the
column which would be contained with in that cluster assignment. One
thing to note, is that if you wish to access columns that are contained
within <a href="mailto:object@sample" class="email">object@sample</a>_metadata.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">marrow_A</span>, </span>
<span>                          reduction <span class="op">=</span> <span class="st">'PCA_UMAP'</span>, </span>
<span>                          assay <span class="op">=</span> <span class="st">'SCT'</span>, </span>
<span>                          clust.method <span class="op">=</span> <span class="st">'PCA_NN:LOUVAIN'</span>, </span>
<span>                          column <span class="op">=</span> <span class="st">'res_0.8'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCT'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">marrow_A</span>, </span>
<span>                          reduction <span class="op">=</span> <span class="st">'PCA_UMAP'</span>, </span>
<span>                          assay <span class="op">=</span> <span class="st">'SCRAN'</span>, </span>
<span>                          clust.method <span class="op">=</span> <span class="st">'PCA_NN:LOUVAIN'</span>, </span>
<span>                          column <span class="op">=</span> <span class="st">'res_0.8'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCRAN'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">marrow_A</span>, </span>
<span>                          reduction <span class="op">=</span> <span class="st">'PCA_UMAP'</span>, </span>
<span>                          assay <span class="op">=</span> <span class="st">'SCANPY'</span>, </span>
<span>                          clust.method <span class="op">=</span> <span class="st">'PCA_NN:LOUVAIN'</span>, </span>
<span>                          column <span class="op">=</span> <span class="st">'res_0.8'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCANPY'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">egg</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/egg/man/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span>plots <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">plot1</span>, <span class="va">plot2</span>, <span class="va">plot3</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="using_IBRAP_files/figure-html/unnamed-chunk-20-1.png" width="700"><img src="using_IBRAP_files/figure-html/unnamed-chunk-21-1.png" width="700"><img src="using_IBRAP_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<p>As we can see, each normalisation method changes the final topology
of the data. Furthermore, we can see that SCT, SCRAN, and SCANPY
produced 19, 15, and 13 communities, respectively. Therefore, it is
important to assess all normalisation methods and their subsequent
results since they can differ and some perform better than others
according to datasets. For this analysis we will proceed using the SCT
results.</p>
</div>
<div class="section level3">
<h3 id="section-10-biological-analysis">Section 10: Biological Analysis<a class="anchor" aria-label="anchor" href="#section-10-biological-analysis"></a>
</h3>
<p>To understand the biology driving the clustering we can tackle it in
two ways: differential expression analyses or visualisation.</p>
<div class="section level4">
<h4 id="section-10-1">Section 10.1<a class="anchor" aria-label="anchor" href="#section-10-1"></a>
</h4>
<p>Differential expression analyses can be performed either 1 cluster
compared against all other cells or 1 cluster against 1 other cluster.
In this tutorial we will use the 1 v all function.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">DE_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.diffexp.all.html">perform.diffexp.all</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">marrow_A</span>, </span>
<span>                              assay <span class="op">=</span> <span class="st">'SCT'</span>, </span>
<span>                              clust.method <span class="op">=</span> <span class="st">'PCA_NN:LOUVAIN'</span>, </span>
<span>                              column <span class="op">=</span> <span class="st">'res_0.8'</span><span class="op">)</span></span></code></pre></div>
<p>Now, lets take a look at the results:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">DE_res</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="op">]</span></span>
<span><span class="co">#&gt;                    p_val avg_log2FC pct.1 pct.2     p_val_adj cluster</span></span>
<span><span class="co">#&gt; VCAN        0.000000e+00   2.139740 0.899 0.066  0.000000e+00       0</span></span>
<span><span class="co">#&gt; CD14        0.000000e+00   1.797290 0.841 0.038  0.000000e+00       0</span></span>
<span><span class="co">#&gt; S100A12    1.590393e-290   3.380330 0.921 0.117 2.387338e-286       0</span></span>
<span><span class="co">#&gt; FCN1       1.003886e-288   2.984599 0.994 0.168 1.506934e-284       0</span></span>
<span><span class="co">#&gt; CSTA       5.805797e-275   2.440009 0.997 0.160 8.715082e-271       0</span></span>
<span><span class="co">#&gt; CTB-61M7.2 2.214807e-271   1.167172 0.640 0.022 3.324647e-267       0</span></span>
<span><span class="co">#&gt; MS4A6A     4.944813e-271   1.488988 0.866 0.081 7.422658e-267       0</span></span>
<span><span class="co">#&gt; MNDA       2.192499e-267   1.699432 0.884 0.097 3.291160e-263       0</span></span>
<span><span class="co">#&gt; CLEC7A     3.196353e-264   1.539125 0.835 0.076 4.798045e-260       0</span></span>
<span><span class="co">#&gt; G0S2       2.205006e-260   2.725010 0.869 0.108 3.309934e-256       0</span></span>
<span><span class="co">#&gt;                  gene</span></span>
<span><span class="co">#&gt; VCAN             VCAN</span></span>
<span><span class="co">#&gt; CD14             CD14</span></span>
<span><span class="co">#&gt; S100A12       S100A12</span></span>
<span><span class="co">#&gt; FCN1             FCN1</span></span>
<span><span class="co">#&gt; CSTA             CSTA</span></span>
<span><span class="co">#&gt; CTB-61M7.2 CTB-61M7.2</span></span>
<span><span class="co">#&gt; MS4A6A         MS4A6A</span></span>
<span><span class="co">#&gt; MNDA             MNDA</span></span>
<span><span class="co">#&gt; CLEC7A         CLEC7A</span></span>
<span><span class="co">#&gt; G0S2             G0S2</span></span></code></pre></div>
<p>These results will produce a table with key columns: the gene names,
pct.1 and pct.2 comparing expression values, and their p-value. These
columns are key in identifying the strength in the difference in
expression and whether the difference is statistically significant.
However, when this fails visualisation can help. We have included 4
different methods of visualisation: feature plots, violin plots, dot
plots, and heatmaps.</p>
<p>In this case, we can use feature plots:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/plot.features.html">plot.features</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">marrow_A</span>, assay <span class="op">=</span> <span class="st">'SCT'</span>, reduction <span class="op">=</span> <span class="st">'PCA_UMAP'</span>, </span>
<span>              features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'JCHAIN'</span>,<span class="st">'MS4A1'</span>,<span class="st">'MME'</span>,<span class="st">'IRF8'</span>,<span class="st">'HLA-DRA'</span>,<span class="st">'LYZ'</span>,</span>
<span>                           <span class="st">'ELANE'</span>,<span class="st">'ALAS2'</span>,<span class="st">'CD34'</span>,<span class="st">'NKG7'</span>,<span class="st">'CCR7'</span>,<span class="st">'IL7R'</span>,<span class="st">'GZMK'</span><span class="op">)</span>, </span>
<span>              pt_size <span class="op">=</span> <span class="fl">1</span>, </span>
<span>              slot <span class="op">=</span> <span class="st">'normalised'</span><span class="op">)</span></span></code></pre></div>
<p><img src="using_IBRAP_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<p>Next, we must indicate which cell type belongs to which clusters. We
can fetch results using our FetchResults function.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">marrow_A</span><span class="op">[[</span><span class="st">'SCT'</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">'PCA_NN:LOUVAIN'</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">'res_0.8'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># here we are replacing the clusters with their cell types</span></span>
<span></span>
<span><span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'18'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'plasma'</span></span>
<span><span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'17'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'immature B cells'</span></span>
<span><span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'11'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'pro B cells'</span></span>
<span><span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'14'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'dendritic cells'</span></span>
<span><span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'15'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'dendritic cells'</span></span>
<span><span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'1'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'monocytes/macrophage'</span></span>
<span><span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'10'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'monocytes/macrophage'</span></span>
<span><span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'6'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'neutrophil'</span></span>
<span><span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'5'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'erythrocytes'</span></span>
<span><span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'8'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'erythrocytes'</span></span>
<span><span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'9'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'erythrocytes'</span></span>
<span><span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'12'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'erythrocytes'</span></span>
<span><span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'16'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'erythrocytes'</span></span>
<span><span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'3'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'HSPCs'</span></span>
<span><span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'4'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'NK cells'</span></span>
<span><span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'0'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'CD4+ T cells'</span></span>
<span><span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'2'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'CD8+ naive T cells'</span></span>
<span><span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'13'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'CD8+ naive T cells'</span></span>
<span><span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'7'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'CD8+ cytotoxic T cells'</span></span>
<span></span>
<span><span class="co"># and we are going to simply put the cell types into our metadata</span></span>
<span></span>
<span><span class="va">marrow_A</span><span class="op">[[</span><span class="st">'celltypes'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">clusters</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">marrow_A</span>, </span>
<span>                 reduction <span class="op">=</span> <span class="st">'PCA_UMAP'</span>, </span>
<span>                 assay <span class="op">=</span> <span class="st">'SCT'</span>, </span>
<span>                 clust.method <span class="op">=</span> <span class="st">'metadata'</span>, </span>
<span>                 column <span class="op">=</span> <span class="st">'celltypes'</span>, </span>
<span>                 pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="using_IBRAP_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
<p>Congratulations for completing your first run through of IBRAP! Thus
far, you have installed a great collection of packages for single-cell
analytics, produced 3 pipelines, benchmarked them, performed a
differential expression, and produced plots. This is just a snippet of
what IBRAP is capable of, feel free to explore the available functions
and investigate the rest of our tutorial website to further elucidate
IBRAP.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Connor Knight, Faraz Khan.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
