<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getting Started • IBRAP</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Getting Started">
<meta property="og:description" content="IBRAP">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">IBRAP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Tutorials
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/installing_ibrap.html">Installation</a>
    </li>
    <li>
      <a href="../articles/using_IBRAP.html">Getting Started</a>
    </li>
    <li>
      <a href="../articles/integration.html">Data Integration</a>
    </li>
    <li>
      <a href="../articles/application.html">RShiny Application</a>
    </li>
    <li>
      <a href="../articles/trajectory.html">Trajectory Inference</a>
    </li>
    <li>
      <a href="../articles/singleR_annotation.html">Automated Annotation</a>
    </li>
    <li>
      <a href="../articles/data_transform.html">Data Transformation</a>
    </li>
    <li>
      <a href="../articles/clustering_method.html">Clustering Methods</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="using_IBRAP_files/clipboard-1.7.1/clipboard.min.js"></script><link href="using_IBRAP_files/primer-tooltips-1.4.0/build.css" rel="stylesheet">
<link href="using_IBRAP_files/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet">
<script src="using_IBRAP_files/klippy-0.0.0.9500/js/klippy.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Getting Started</h1>
                        <h4 data-toc-skip class="author">Connor H Knight &amp; Faraz Khan</h4>
            
            <h4 data-toc-skip class="date">11/08/2021</h4>
      
      
      <div class="hidden name"><code>using_IBRAP.Rmd</code></div>

    </div>

    
    
<script>
  addClassKlippyTo("pre.r, pre.markdown");
  addKlippy('right', 'bottom', 'auto', '1', 'Click to copy', 'Done');
</script><div id="analysing-2994-bone-marrow-mononuclear-cells-bmmc-from-a-single-sample-" class="section level1">
<h1 class="hasAnchor">
<a href="#analysing-2994-bone-marrow-mononuclear-cells-bmmc-from-a-single-sample-" class="anchor" aria-hidden="true"></a>Analysing 2994 Bone Marrow Mononuclear Cells (BMMC) from a single sample.</h1>
<p>To demonstrate IBRAP we enlisted a sample produced by <a href="https://insight.jci.org/articles/view/124928" class="external-link">Oetjen</a> <em>et al</em>. This can be downloaded using the following link:</p>
<p><a href="https://www.dropbox.com/s/s126bv59mjnhkre/marrow_A.rds?dl=0" class="external-link uri">https://www.dropbox.com/s/s126bv59mjnhkre/marrow_A.rds?dl=0</a></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.choose.html" class="external-link">file.choose</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
  </code></pre></div>
<div id="section-1-cleaning-droplet-based-datasdets" class="section level2">
<h2 class="hasAnchor">
<a href="#section-1-cleaning-droplet-based-datasdets" class="anchor" aria-hidden="true"></a>Section 1: Cleaning droplet-based datasdets</h2>
<div id="section-1-1-applying-the-functions" class="section level3">
<h3 class="hasAnchor">
<a href="#section-1-1-applying-the-functions" class="anchor" aria-hidden="true"></a>Section 1.1: Applying the functions</h3>
<p>Droplet-based datasets present unique confounding effects, including: doublets and ambient RNA contamination. To remove them we have enlisted scrublet and decontX, respectively. Thus far, there is no consensus on the order of application. However, we prefer scrublet followed by decontX.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># We simply supply our generated count matrix to both functions</span>

<span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.scrublet.html">perform.scrublet</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">bmmc</span><span class="op">)</span>

<span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.decontX.html">perform.decontX</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">bmmc</span><span class="op">)</span></code></pre></div>
</div>
<div id="section-1-2-explaining-the-results" class="section level3">
<h3 class="hasAnchor">
<a href="#section-1-2-explaining-the-results" class="anchor" aria-hidden="true"></a>Section 1.2: Explaining the results</h3>
<p><img src="scrublet_plot.png" width="150%" height="150%"></p>
<p>Scrublet functions by simulating multiplets from out datasets (by integrating similar cells); we can see these simulated cells with high doublet scores. For this example, we can see a minimal number of cells in our observed cells with high doublet scores. Cells that are above a certain doublet score threshold are omitted from the dataset.</p>
<p><img src="decontX.png" width="150%" height="150%"></p>
<p>DecontX creates a umap projection of samples and colours them according the amount of contamination presented within a cell. In this sample, we observe minimal contaminaton. WARNING: DecontX should only be used if you expect ambient RNA otherwise it can drastically affect downstream results.</p>
<p>Now that we have removed these confounding effects from our dataset we can move on with our analysis.</p>
</div>
</div>
<div id="section-2-initiating-the-ibrap-s4-object" class="section level2">
<h2 class="hasAnchor">
<a href="#section-2-initiating-the-ibrap-s4-object" class="anchor" aria-hidden="true"></a>Section 2: Initiating the IBRAP S4 object</h2>
<p>Lets start by initiating our IBRAP class object ready for downstream analyses. You can see the structure and a manual for these in our reference section.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># whilst creating the object we begin filtering the cells and genes using min.cells and min.features</span>
<span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createIBRAPobject.html">createIBRAPobject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">bmmc</span>,
                          original.project <span class="op">=</span> <span class="st">'bmmc'</span>,
                          method.name <span class="op">=</span> <span class="st">'RAW'</span>,
                          min.cells <span class="op">=</span> <span class="fl">3</span>,
                          min.features <span class="op">=</span> <span class="fl">200</span><span class="op">)</span>
                          </code></pre></div>
<p>This produces an S4 class object that initially contains sample metadata and a raw method assay contained in <a href="mailto:object@methods" class="email">object@methods</a>. A couple of important steps happen here: metadata, including: total counts and total genes per cell are stored in <a href="mailto:object@sample_metadata" class="email">object@sample_metadata</a>, and some initial filtering of the count matrix happens. In this case, we removed any genes that are not present in at least 3 cells and removed cells with less than 200 genes We performed the latter as to remove any droplets that are empty and do not contain true cells.</p>
</div>
<div id="section-3-quality-control" class="section level2">
<h2 class="hasAnchor">
<a href="#section-3-quality-control" class="anchor" aria-hidden="true"></a>Section 3: Quality Control</h2>
<div id="section-3-1-metrices" class="section level3">
<h3 class="hasAnchor">
<a href="#section-3-1-metrices" class="anchor" aria-hidden="true"></a>Section 3.1: Metrices</h3>
<p>Next, we must assess the quality of our datasets, if you wish to read in-depth about Quality Control (QC) then please refer to this <a href="https://www.embopress.org/doi/full/10.15252/msb.20188746" class="external-link">publication</a>. In summary, we use the following metrices to QC our data:</p>
<table class="table">
<colgroup>
<col width="12%">
<col width="87%">
</colgroup>
<thead><tr class="header">
<th align="left">QC.Metric</th>
<th align="left">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Mitochondrial Fraction</td>
<td align="left">An abnormally high mitochondrail fraction indicates improper wet lab preparation</td>
</tr>
<tr class="even">
<td align="left">Total RNA</td>
<td align="left">A difference in count value can affect downstream processes and requires normalising downstream</td>
</tr>
<tr class="odd">
<td align="left">Total Genes</td>
<td align="left">Low gene presence indicates droplets containing no cells whereas, disproportionately high number of genes indicates a droplet containing more than one cell</td>
</tr>
</tbody>
</table>
<p>As previously mentioned we already generated total count and total genes whilst creating our ibrap object. However, percentage mitochondrial genes requires a second function since mitochondrial genes have a varying prefix in datasets:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_percentage_genes.html">find_percentage_genes</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">bmmc</span>, pattern <span class="op">=</span> <span class="st">'^MT-'</span>,
                              assay <span class="op">=</span> <span class="st">'RAW'</span>, 
                              slot <span class="op">=</span> <span class="st">'counts'</span>,
                              column.name <span class="op">=</span> <span class="st">'RAW_percent.mt'</span><span class="op">)</span>
                              </code></pre></div>
<p>These fraction results will be saved under <a href="mailto:object@sample_metadata" class="email">object@sample_metadata</a> alongside total counts and total genes for use downstream.</p>
</div>
<div id="section-3-2-visualisation" class="section level3">
<h3 class="hasAnchor">
<a href="#section-3-2-visualisation" class="anchor" aria-hidden="true"></a>Section 3.2: Visualisation</h3>
<p>Visualising this information is important for high dimensional data. Thus, we can achieve this in two ways: using violin plots or comparing two metrices with scatter plots.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="fu"><a href="../reference/plot.QC.vln.html">plot.QC.vln</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">bmmc</span>, 
            metadata.columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'RAW_total.features'</span>, 
                                 <span class="st">'RAW_total.counts'</span>, 
                                 <span class="st">'RAW_percent.mt'</span><span class="op">)</span><span class="op">)</span>
                                 </code></pre></div>
<p>The violin plot is split into its metrices and displays the volume of cells expressing the value on the y axis.</p>
<p><img src="vln_qc_pre.png" width="150%" height="150%"></p>
<p>Scatter plots enable us to observe correlations between metrices, i.e. the higher the amount of features in a cell the higher the number of counts there should be.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="fu">plot.qc.scatter</span><span class="op">(</span>object <span class="op">=</span> <span class="va">bmmc</span>, 
                x <span class="op">=</span> <span class="st">'RAW_total.counts'</span>, 
                y <span class="op">=</span> <span class="st">'RAW_total.features'</span>, 
                split.by <span class="op">=</span> <span class="st">'original.project'</span>
                <span class="op">)</span></code></pre></div>
<p><img src="scatter_qc_pre.png" width="150%" height="150%"></p>
<p>Now that we have eye-balled our data we can proceed to filtration.</p>
</div>
<div id="section-3-3-filtration" class="section level3">
<h3 class="hasAnchor">
<a href="#section-3-3-filtration" class="anchor" aria-hidden="true"></a>Section 3.3: Filtration</h3>
<p>We clean our data further by applyng filtration cut-offs. We remove cells with a high number of features since this will remove any doublets that were missed by scrublet (two cells in a droplet will display a disproportionately higher number of features). Then, we remove cells with a certain fraction of mitochondrial counts, we do this to remove cells with damaged mitochondria during the preparation process.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_IBRAP.html">filter_IBRAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">bmmc</span>, <span class="va">RAW_total.features</span> <span class="op">&lt;</span> <span class="fl">3000</span> <span class="op">&amp;</span> <span class="va">RAW_percent.mt</span> <span class="op">&lt;</span> <span class="fl">8</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="section-4-data-transformation" class="section level2">
<h2 class="hasAnchor">
<a href="#section-4-data-transformation" class="anchor" aria-hidden="true"></a>Section 4: Data Transformation</h2>
<div id="section-4-1-applying-the-functions" class="section level3">
<h3 class="hasAnchor">
<a href="#section-4-1-applying-the-functions" class="anchor" aria-hidden="true"></a>Section 4.1: Applying The Functions</h3>
<p>To accommodate downstream analyses, we must promote a normal distribution and numerical scaling in our data. We included four separate <a href="https://connorhknight.github.io/IBRAP/articles/data_transform.html" class="external-link">normalisation</a> methods that are applied as such:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"></a>
<a class="sourceLine" id="cb8-2" data-line-number="2">bmmc &lt;-<span class="st"> </span><span class="kw">perform.sct</span>(<span class="dt">object =</span> bmmc, </a>
<a class="sourceLine" id="cb8-3" data-line-number="3">                    <span class="dt">assay =</span> <span class="st">'RAW'</span>, </a>
<a class="sourceLine" id="cb8-4" data-line-number="4">                    <span class="dt">slot =</span> <span class="st">'counts'</span>, <span class="dt">vars.to.regress =</span> <span class="st">'RAW_percent.mt'</span>)</a>
<a class="sourceLine" id="cb8-5" data-line-number="5"></a>
<a class="sourceLine" id="cb8-6" data-line-number="6">bmmc &lt;-<span class="st"> </span><span class="kw">perform.scran</span>(<span class="dt">object =</span> bmmc, </a>
<a class="sourceLine" id="cb8-7" data-line-number="7">                      <span class="dt">assay =</span> <span class="st">'RAW'</span>, </a>
<a class="sourceLine" id="cb8-8" data-line-number="8">                      <span class="dt">slot =</span> <span class="st">'counts'</span>, </a>
<a class="sourceLine" id="cb8-9" data-line-number="9">                      <span class="dt">vars.to.regress =</span> <span class="kw">c</span>(<span class="st">'RAW_percent.mt'</span> <span class="st">'RAW_total.counts'</span>))</a>
<a class="sourceLine" id="cb8-10" data-line-number="10"></a>
<a class="sourceLine" id="cb8-11" data-line-number="11">bmmc &lt;-<span class="st"> </span><span class="kw">perform.scanpy</span>(<span class="dt">object =</span> bmmc, </a>
<a class="sourceLine" id="cb8-12" data-line-number="12">                       <span class="dt">vars.to.regress =</span> <span class="kw">c</span>(<span class="st">'RAW_percent.mt'</span> <span class="st">'RAW_total.counts'</span>))</a></code></pre></div>
<p>For these methods, we transformed our raw counts, and regressed some QC metrices from our datasets to not allow them to affect our downstream analyses. There are some notable differences here. For example, SCT automatically regresses RAW_total.counts so we do not need to specify this however, the others you do. These parameters are set as default, any deviation from these is at your own risk since they could produce incorrect results. Nonetheless, feel free to play around and observe the difference.</p>
</div>
<div id="section-4-2-how-functions-work-downstream" class="section level3">
<h3 class="hasAnchor">
<a href="#section-4-2-how-functions-work-downstream" class="anchor" aria-hidden="true"></a>Section 4.2: How Functions Work Downstream</h3>
<p>Each normalisation function produces a new method-assay which contains the results. Any downstream analyses will be stored in these separate compartments, respective to the normalisation method. i.e. <a href="mailto:object@methods" class="email">object@methods</a> contains all of the method assays. Hereon, function will iterate through the indicated method-assays and any other inputs within the assays. For example, if you supply c(‘SCT’,‘SCRAN’,‘SCANPY’) as assays and c(‘pca_umap’, ‘pca_t-SNE’) as reductions. A function will perform firstly on ‘pca_umap’ in the ‘SCT’ assay, then the ‘pca_t-SNE’ in the ‘SCT’ assay. After these the function will move on to ‘pca_umap’ in the ‘SCRAN’ assay and so on.</p>
</div>
<div id="section-4-3-downstream-pipeline-deviations" class="section level3">
<h3 class="hasAnchor">
<a href="#section-4-3-downstream-pipeline-deviations" class="anchor" aria-hidden="true"></a>Section 4.3: Downstream Pipeline Deviations</h3>
<p>Since IBRAP incorporated multiple pipelines some pipelines begin to vary at this point. For example, Scanpy PCA calculation and neighbourhood graph generation slightly differ. IBRAP accomodates this by identifyig the assay name and automatically applying the correct function. When adding suffixes to any result naming it is important not to include your own underscores.</p>
</div>
</div>
<div id="section-5-removing-certain-highly-variable-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#section-5-removing-certain-highly-variable-genes" class="anchor" aria-hidden="true"></a>Section 5: Removing Certain Highly Variable Genes</h2>
<p>In some instances, certain genes require omission from the highly varaible genes. This can be achieved with the following function:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># This function finds all ribosomal genes and removes them from the HVG list of the three assays</span>
<span class="va">unwanted</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="../reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">bmmc</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^RPL"</span> ,x<span class="op">=</span><span class="fu"><a href="../reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">bmmc</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remove.hvgs.html">remove.hvgs</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">bmmc</span>, assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, 
                    hvgs.omit <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">unwanted</span>,<span class="va">unwanted</span>,<span class="va">unwanted</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="section-6-linear-dimensionaly-reduction" class="section level2">
<h2 class="hasAnchor">
<a href="#section-6-linear-dimensionaly-reduction" class="anchor" aria-hidden="true"></a>Section 6: Linear Dimensionaly Reduction</h2>
<div id="section-6-1-principal-component-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#section-6-1-principal-component-analysis" class="anchor" aria-hidden="true"></a>Section 6.1: Principal Component Analysis</h3>
<p>Now that our data has been appropriately transformed and highly varaible genes have been identified we can reduce the number of dimensions using Principal Component Analysis (PCA). Reduction methods summarise as much information from the higher dimensions into lower dimensions as possible. Therefore, these lower dimensions facilitate a faster computing time and require less computational resources.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.pca.html">perform.pca</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">bmmc</span>, 
                    assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, 
                    n.pcs <span class="op">=</span> <span class="fl">50</span>, reduction.save <span class="op">=</span> <span class="st">'pca'</span><span class="op">)</span></code></pre></div>
<p>As you can see, we have requested this function to iterate through all three assays and to save the reduction as pca. The produced principal components (PCs) contain a different explained variance, high variance PCs contain the most information between points and low provide the opposite. Furthermore, low variation PCs create a false closeness between cell points and thus require exclusion from <em>some</em> downstream analyses. A simple way to identify the number of PCs to retain is to plot them in descending order and find the elbow point. The elbow point being the bend in the plot of points.</p>
<p><img src="PCA_screeplot.png" width="150%" height="150%"></p>
<p>For these assays, the elbow point is around 10-15 PCs.</p>
</div>
<div id="section-6-2-diffusion-maps" class="section level3">
<h3 class="hasAnchor">
<a href="#section-6-2-diffusion-maps" class="anchor" aria-hidden="true"></a>Section 6.2: Diffusion Maps</h3>
<p>We use next calculate diffusion maps from our PCA reduction. Diffusion maps are useful in visualising the developmental trajectory in our cellular population once processed with a non-linear method, such as: UMAP. To create this we apply the following function:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.diffusion.map.html">perform.diffusion.map</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">bmmc</span>, assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, reduction <span class="op">=</span> <span class="st">'pca'</span>, dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Again, we have requested for IBRAP to iterate through our three assays whilst using their PCA reduction. We specified the dimensions to use as 0 (which means all components) and the default setting for this function produces 15 diffusion components. Both the PCA and diffusion map will be saved in the method assay in <a href="mailto:assay@computational_reductions" class="email">assay@computational_reductions</a>.</p>
<p>We will then process these downstream. Next, we will cluster our datasets.</p>
</div>
</div>
<div id="section-7-generating-neighbourhood-graphs-clustering" class="section level2">
<h2 class="hasAnchor">
<a href="#section-7-generating-neighbourhood-graphs-clustering" class="anchor" aria-hidden="true"></a>Section 7: Generating Neighbourhood Graphs &amp; Clustering</h2>
<div id="section-7-1-neighbourhood-graphs" class="section level3">
<h3 class="hasAnchor">
<a href="#section-7-1-neighbourhood-graphs" class="anchor" aria-hidden="true"></a>Section 7.1: Neighbourhood Graphs</h3>
<p>Prior to clustering it is important to ascertain the relationship and closeness of our cells. As a pre-requisite to clustering we are going to produce a k-nearest neghbour graphs. SCT, SCRAN, and SCANPY use one algorithm whilst SCANPY uses its own version. This is produced by applying the following function:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># here we generate the neighbour graph for PCA and produce a diffusion map from the PCA-neighbour graph</span>

<span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.nn.html">perform.nn</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">bmmc</span>, assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, 
                   reduction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'pca'</span>, <span class="st">'pca:diffmap'</span><span class="op">)</span>, 
                   dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span>
                          </code></pre></div>
<p>In this function, we have requested that the function iterate through all three assays, within each assay to use pca and then pca:diffmap using all dimensions (seen as list(0,0)). These results will be stored under <a href="mailto:assay@neighbours" class="email">assay@neighbours</a> ready for use downstream.</p>
</div>
<div id="section-7-2-clustering" class="section level3">
<h3 class="hasAnchor">
<a href="#section-7-2-clustering" class="anchor" aria-hidden="true"></a>Section 7.2: Clustering</h3>
<p>Finally, using our neighbourhood graph we must cluster close networks together to infer individual cell types. perform.graph.cluster() provides either: - Louvain algorithm (1) - Louvain algorithm with multilevel refinement (2) - Smart Local Moving (SLM) (3) - Leiden (4)</p>
<p>These methods can be adjusted using thre res parameter in the function typically ranging between 0.1-1.5. A higher resolution finds more communities in a sample, whereas a lower value produces the opposite. The default option for this function is to use the following resolutions: c(0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1, 1.1, 1.2, 1.3, 1.4, 1.5). However, feel free to adjsut this to your requirement.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.graph.cluster.html">perform.graph.cluster</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">bmmc</span>, 
                              assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, 
                              neighbours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pca_nn"</span>, <span class="st">"pca:diffmap_nn"</span><span class="op">)</span>, 
                              algorithm <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>For this example we clustered all assays and used our neighbourhood graphs generated from PCA and Diffusion maps using the classic louvain algorithm. As mentioned above there are a few graph-based methods. But, we have also inclusing a couple of <a href="https://connorhknight.github.io/IBRAP/articles/clustering_method.html" class="external-link">alternatives</a> that are described in another tutorial</p>
</div>
</div>
<div id="section-8-non-linear-reduction" class="section level2">
<h2 class="hasAnchor">
<a href="#section-8-non-linear-reduction" class="anchor" aria-hidden="true"></a>Section 8: Non-linear Reduction</h2>
<p>Now we have reduced the dimensionalty of our datasets using non-linear methods and clustered the cells according to their similarity we must visualise these results. However, linear methods such as PCA and diffusion maps are inherently innapropriate for visualisation. Therefore, we must process then using non-linear methods.</p>
<p>IBRAP integrated the ability to perform 3 methods: - UMAP - t-SNE - lvish (a LargeVis adaptation)</p>
<p>For this tutorial we will focus on UMAP.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.umap.html">perform.umap</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">bmmc</span>, 
                     assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span><span class="op">)</span>, 
                     reduction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'pca'</span>, <span class="st">'pca:diffmap'</span><span class="op">)</span>, 
                     n.dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">15</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Scanpy UMAP reductions are produced from the nearest neighbourhood graph that we previously preoduced using the perform.nn() function previously. SCANPY would therefore be processed like this:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.umap.html">perform.umap</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">bmmc</span>, 
                     assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCANPY'</span><span class="op">)</span>, 
                     graph <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'pca_nn'</span><span class="op">)</span>, 
                     n.dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">15</span><span class="op">)</span><span class="op">)</span>

<span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.umap.html">perform.umap</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">bmmc</span>, 
                     assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCANPY'</span><span class="op">)</span>, 
                     reduction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'pca:diffmap'</span><span class="op">)</span>, 
                     n.dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Simply, we asked this function to iterate through all assays, and use the first 15 components of the pca reduction and to use all components of the diffusion map.</p>
<p>Now, we have generated a high volume of results, especially clustering results. Therefore, it is beneficial to applying to some benchmarking comparisons.</p>
</div>
<div id="section-9-benchmarking-results" class="section level2">
<h2 class="hasAnchor">
<a href="#section-9-benchmarking-results" class="anchor" aria-hidden="true"></a>Section 9: Benchmarking Results</h2>
<p>IBRAP contains a benchmarking function that enlists multiple metrices to assess clustering performance. These metrices are only informative however and will be unable to distinguish the best results (that is down to the user to decide).</p>
<table class="table">
<colgroup>
<col width="16%">
<col width="83%">
</colgroup>
<thead><tr class="header">
<th align="left">QC.Metric</th>
<th align="left">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Average Silhouette Width (ASW)</td>
<td align="left">This measures how close cells are to their own cluster (cohesion) whilst comparing to other clusters (separation) A higher value is good.</td>
</tr>
<tr class="even">
<td align="left">Dunn Index</td>
<td align="left">Comapres the ratio of the smallest distance between observations not in the same cluster to the largest intra-cluster distance. A higher value is preferred.</td>
</tr>
<tr class="odd">
<td align="left">Connectivity</td>
<td align="left">This uses a k-nearest neighbour algorithm that determines the connectedness between clusters. A depleted value is prefered.</td>
</tr>
</tbody>
</table>
<p>To set up the function, supply the dataframes contained within the <a href="mailto:assay@cluster_assignments" class="email">assay@cluster_assignments</a> you wish to benchmark under clustering. Next, include the corresponding reductions for them, normally this will be a umap reduction using 1:2 dimensions. For example, if clustering and the umap were calculated using PCA these should be paired, likewise if they were calcualted using pca:diffmap then they should be paired.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">bmmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/benchmark.clustering.html">benchmark.clustering</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">bmmc</span>, 
                             assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SCT'</span>, <span class="st">'SCRAN'</span>, <span class="st">'SCANPY'</span><span class="op">)</span>, 
                             clustering <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'pca_nn:louvain'</span>, <span class="st">'pca:diffmap_nn:louvain'</span><span class="op">)</span>, 
                             reduction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'pca_umap'</span>, <span class="st">'pca:diffmap_umap'</span><span class="op">)</span>, 
                             n.dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span></code></pre></div>
<p>We can visualise the comparison within a clustering method dataframe using the following function:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="fu">plot.cluster.benchmarking</span><span class="op">(</span>object <span class="op">=</span> <span class="va">bmmc</span>, assay <span class="op">=</span> <span class="st">'SCT'</span>, 
                          clustering <span class="op">=</span> <span class="st">'pca_nn:louvain'</span><span class="op">)</span></code></pre></div>
<p><img src="cluster_benchmarking.png" width="150%" height="150%"></p>
<p>This provides a line graph showing the metrices. As you can see, there is a correlation between the results. However, do not be fooled by the preferred valuei n very low settings, this is likely due to a smaller number of communities being detected at lower resolutions which means that clusters are typically more separated.</p>
<p>Now that we have our results and benchmarking to guide us we can begin to visualise the results.</p>
</div>
<div id="section-10-visualising-clustering-results" class="section level2">
<h2 class="hasAnchor">
<a href="#section-10-visualising-clustering-results" class="anchor" aria-hidden="true"></a>Section 10: Visualising Clustering Results</h2>
<p>Since we have 3 different normalisation method it is beneficial to observe the difference in their produces topology and clustering.</p>
<p>The main challenge with the plotting function in IBRAP is isolating the correct labelling. In general, plotting functions will ask for the assay which is contained within <a href="mailto:object@methods" class="email">object@methods</a>, the clust.method that is contained withing <span class="citation">@cluster_assignments</span>, and the column which would be contained with in that cluster assignment. One thing to note, is that if you wish to access columns that are contained within <a href="mailto:object@sample_metadata" class="email">object@sample_metadata</a>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">bmmc</span>, 
                          reduction <span class="op">=</span> <span class="st">'pca_umap'</span>, 
                          assay <span class="op">=</span> <span class="st">'SCT'</span>, 
                          clust.method <span class="op">=</span> <span class="st">'pca_nn:louvain'</span>, 
                          column <span class="op">=</span> <span class="st">'neighbourhood_graph_res.0.8'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCT'</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span>

<span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">bmmc</span>, 
                          reduction <span class="op">=</span> <span class="st">'pca_umap'</span>, 
                          assay <span class="op">=</span> <span class="st">'SCRAN'</span>, 
                          clust.method <span class="op">=</span> <span class="st">'pca_nn:louvain'</span>, 
                          column <span class="op">=</span> <span class="st">'neighbourhood_graph_res.0.8'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCRAN'</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span>

<span class="va">plot3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">bmmc</span>, 
                          reduction <span class="op">=</span> <span class="st">'pca_umap'</span>, 
                          assay <span class="op">=</span> <span class="st">'SCANPY'</span>, 
                          clust.method <span class="op">=</span> <span class="st">'pca_nn:louvain'</span>, 
                          column <span class="op">=</span> <span class="st">'neighbourhood_graph_res.0.8'</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">'SCANPY'</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggplot2/man/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span>

<span class="va">plot1</span> <span class="op">+</span> <span class="va">plot2</span> <span class="op">+</span> <span class="va">plot3</span></code></pre></div>
<p><img src="diff_norm_method.png" width="150%" height="150%"></p>
<p>As we can see, each normalisation method changes the final topology of the data. Furthermore, we can see that SCT, SCRAN, and SCANPY produced 18, 14, and 14 communities, respectively. Therefore, it is important to assess all normalisation methods and their subsequent results since they can differ and some perform better than others according to datasets. For this analysis we will proceed using the SCT results.</p>
</div>
<div id="section-11-biological-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#section-11-biological-analysis" class="anchor" aria-hidden="true"></a>Section 11: Biological Analysis</h2>
<p>To understand the biology driving the clustering we can tackle it in two ways: differential expression analyses or visualisation.</p>
<div id="section-11-1" class="section level3">
<h3 class="hasAnchor">
<a href="#section-11-1" class="anchor" aria-hidden="true"></a>Section 11.1</h3>
<p>Differential expression analyses can be performed either 1 cluster compared against all other cells or 1 cluster against 1 other cluster. In this tutorial we will use the 1 v all function.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">DE_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perform.diffexp.all.html">perform.diffexp.all</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">samp</span>, 
                              assay <span class="op">=</span> <span class="st">'SCT'</span>, 
                              test <span class="op">=</span> <span class="st">'MAST'</span>, 
                              identity <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">samp</span><span class="op">@</span><span class="va">methods</span><span class="op">$</span><span class="va">SCT</span><span class="op">@</span><span class="va">cluster_assignments</span><span class="op">$</span><span class="va">`pca_nn.v1:louvain`</span><span class="op">$</span><span class="va">neighbourhood_graph_res.0.8</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>These results will produce a table with key columns: the gene names, pct.1 and pct.2 comparing expression values, and their p-value. These columns are key in identifying the strength in the difference in expression and whether the difference is statistically significant. However, when this fails visualisation can help. We have included 4 different methods of visualisation: feature plots, violin plots, dot plots, and heatmaps.</p>
<p>In this case, we can use feature plots:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="fu"><a href="../reference/plot.features.html">plot.features</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">samp</span>, assay <span class="op">=</span> <span class="st">'SCT'</span>, reduction <span class="op">=</span> <span class="st">'pca_umap'</span>, 
              features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'JCHAIN'</span>,<span class="st">'MS4A1'</span>,<span class="st">'MME'</span>,<span class="st">'IRF8'</span>,<span class="st">'HLA-DRA'</span>,<span class="st">'LYZ'</span>,
                           <span class="st">'ELANE'</span>,<span class="st">'ALAS2'</span>,<span class="st">'CD34'</span>,<span class="st">'NKG7'</span>,<span class="st">'CCR7'</span>,<span class="st">'IL7R'</span>,<span class="st">'GZMK'</span><span class="op">)</span>, 
              pt_size <span class="op">=</span> <span class="fl">1</span>, 
              slot <span class="op">=</span> <span class="st">'normalised'</span><span class="op">)</span></code></pre></div>
<p><img src="Feature_plots.png" width="150%" height="150%"></p>
<p>Next, we must indicate which cell type belongs to which clusters</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">bmmc</span><span class="op">@</span><span class="va">methods</span><span class="op">$</span><span class="va">SCT</span><span class="op">@</span><span class="va">cluster_assignments</span><span class="op">$</span><span class="va">`pca_nn:louvain`</span><span class="op">$</span><span class="va">neighbourhood_graph_res.0.8</span><span class="op">)</span>

<span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'18'</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'plasma'</span>
<span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'17'</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'immature B cells'</span>
<span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'11'</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'pro B cells'</span>
<span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'14'</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'dendritic cells'</span>
<span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'15'</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'dendritic cells'</span>
<span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'1'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'monocytes/macrophage'</span>
<span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'10'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'monocytes/macrophage'</span>
<span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'6'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'neutrophil'</span>
<span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'5'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'erythrocytes'</span>
<span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'8'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'erythrocytes'</span>
<span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'9'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'erythrocytes'</span>
<span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'12'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'erythrocytes'</span>
<span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'16'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'erythrocytes'</span>
<span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'3'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'HSPCs'</span>
<span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'4'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'NK cells'</span>
<span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'0'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'CD4+ T cells'</span>
<span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'2'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'CD8+ naive T cells'</span>
<span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'13'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'CD8+ naive T cells'</span>
<span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="st">'7'</span><span class="op">]</span> <span class="op">=</span> <span class="st">'CD8+ cytotoxic T cells'</span>

<span class="va">bmmc</span><span class="op">@</span><span class="va">sample_metadata</span><span class="op">$</span><span class="va">celltypes</span> <span class="op">&lt;-</span> <span class="va">clusters</span>

<span class="fu"><a href="../reference/plot.reduced.dim.html">plot.reduced.dim</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">bmmc</span>, 
                 reduction <span class="op">=</span> <span class="st">'pca_umap'</span>, 
                 assay <span class="op">=</span> <span class="st">'SCT'</span>, 
                 clust.method <span class="op">=</span> <span class="st">'metadata'</span>, 
                 column <span class="op">=</span> <span class="st">'celltypes'</span>, 
                 pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></code></pre></div>
<p><img src="cell_types.png" width="150%" height="150%"></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Connor Knight, Faraz Khan.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link external-link">pkgdown</a> 1.6.1.9001.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
